{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b09ec3a0",
   "metadata": {},
   "source": [
    "This file describes how we performed in-silico perturbation of TFs. There are two types of perturbation:\n",
    "\n",
    "1. in-silico knockout: zero-out TF expressions\n",
    "2. in-silico upregulation: for a TF, add 0.9-quantile of its expression values across all samples to its original expression values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bd0d889",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pyfaidx import Fasta\n",
    "from verstack import stratified_continuous_split\n",
    "import matplotlib.pyplot as plt\n",
    "import time\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import average_precision_score\n",
    "from sklearn.preprocessing import StandardScaler, MinMaxScaler\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.metrics import mean_squared_error\n",
    "from sklearn.metrics import f1_score\n",
    "from sklearn import preprocessing\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "import scipy\n",
    "from scipy import stats\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "import torch.optim as optim\n",
    "from torch import Tensor\n",
    "from torch.utils.data import TensorDataset, DataLoader\n",
    "from torch.optim.lr_scheduler import StepLR\n",
    "from torch.utils.data import Dataset\n",
    "from torchvision import datasets\n",
    "from torcheval.metrics.functional import multiclass_f1_score\n",
    "\n",
    "import h5py\n",
    "import seaborn as sns\n",
    "import os\n",
    "os.chdir('/nfs/public/xixi/scRegulate/SHAREseq')\n",
    "\n",
    "from typing import Tuple"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea1ae3cb",
   "metadata": {},
   "source": [
    "# Data and model preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "950b3dc7",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr18-31634804-31635104</th>\n",
       "      <th>chr1-165460961-165461261</th>\n",
       "      <th>chr1-75317415-75317715</th>\n",
       "      <th>chr18-46525819-46526119</th>\n",
       "      <th>chr4-3938558-3938858</th>\n",
       "      <th>chr6-71440583-71440883</th>\n",
       "      <th>chr7-105399959-105400259</th>\n",
       "      <th>chr1-161876704-161877004</th>\n",
       "      <th>chr6-113306525-113306825</th>\n",
       "      <th>chr1-177796516-177796816</th>\n",
       "      <th>...</th>\n",
       "      <th>chr10-111689260-111689560</th>\n",
       "      <th>chr10-111388777-111389077</th>\n",
       "      <th>chr10-110841761-110842061</th>\n",
       "      <th>chr10-10884983-10885283</th>\n",
       "      <th>chr10-108335612-108335912</th>\n",
       "      <th>chr10-107887768-107888068</th>\n",
       "      <th>chr10-107038880-107039180</th>\n",
       "      <th>chr10-10549625-10549925</th>\n",
       "      <th>chr10-105270865-105271165</th>\n",
       "      <th>chr10-103291333-103291633</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>V1</th>\n",
       "      <td>2.112241</td>\n",
       "      <td>1.988862</td>\n",
       "      <td>1.988862</td>\n",
       "      <td>1.664337</td>\n",
       "      <td>1.758608</td>\n",
       "      <td>2.112241</td>\n",
       "      <td>1.758608</td>\n",
       "      <td>1.434083</td>\n",
       "      <td>1.919221</td>\n",
       "      <td>1.919221</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.554779</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V2</th>\n",
       "      <td>2.184422</td>\n",
       "      <td>2.079218</td>\n",
       "      <td>1.957768</td>\n",
       "      <td>1.533112</td>\n",
       "      <td>1.814123</td>\n",
       "      <td>1.889215</td>\n",
       "      <td>1.957768</td>\n",
       "      <td>1.814123</td>\n",
       "      <td>1.731114</td>\n",
       "      <td>1.814123</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.546106</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V3</th>\n",
       "      <td>2.381858</td>\n",
       "      <td>1.865213</td>\n",
       "      <td>1.717455</td>\n",
       "      <td>1.542694</td>\n",
       "      <td>2.298454</td>\n",
       "      <td>1.542694</td>\n",
       "      <td>2.298454</td>\n",
       "      <td>1.993207</td>\n",
       "      <td>1.865213</td>\n",
       "      <td>1.865213</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V4</th>\n",
       "      <td>2.296301</td>\n",
       "      <td>1.952485</td>\n",
       "      <td>1.863144</td>\n",
       "      <td>1.952485</td>\n",
       "      <td>1.364729</td>\n",
       "      <td>2.107085</td>\n",
       "      <td>1.763270</td>\n",
       "      <td>1.952485</td>\n",
       "      <td>1.863144</td>\n",
       "      <td>2.174958</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V5</th>\n",
       "      <td>2.021073</td>\n",
       "      <td>1.682703</td>\n",
       "      <td>1.763392</td>\n",
       "      <td>1.903019</td>\n",
       "      <td>1.763392</td>\n",
       "      <td>1.836383</td>\n",
       "      <td>1.682703</td>\n",
       "      <td>1.763392</td>\n",
       "      <td>1.490239</td>\n",
       "      <td>1.490239</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.841352</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V839</th>\n",
       "      <td>3.323174</td>\n",
       "      <td>3.171114</td>\n",
       "      <td>3.171114</td>\n",
       "      <td>2.585932</td>\n",
       "      <td>2.322799</td>\n",
       "      <td>2.585932</td>\n",
       "      <td>3.323174</td>\n",
       "      <td>2.585932</td>\n",
       "      <td>2.000750</td>\n",
       "      <td>2.808408</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V840</th>\n",
       "      <td>2.806491</td>\n",
       "      <td>2.806491</td>\n",
       "      <td>2.055718</td>\n",
       "      <td>2.656048</td>\n",
       "      <td>2.656048</td>\n",
       "      <td>2.485490</td>\n",
       "      <td>1.770699</td>\n",
       "      <td>2.055718</td>\n",
       "      <td>2.806491</td>\n",
       "      <td>2.288595</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V841</th>\n",
       "      <td>2.915099</td>\n",
       "      <td>1.888067</td>\n",
       "      <td>1.626292</td>\n",
       "      <td>2.101953</td>\n",
       "      <td>2.577613</td>\n",
       "      <td>1.888067</td>\n",
       "      <td>2.282790</td>\n",
       "      <td>1.626292</td>\n",
       "      <td>2.439439</td>\n",
       "      <td>2.282790</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.813146</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V842</th>\n",
       "      <td>1.385682</td>\n",
       "      <td>1.421908</td>\n",
       "      <td>1.021733</td>\n",
       "      <td>1.209012</td>\n",
       "      <td>1.346771</td>\n",
       "      <td>1.385682</td>\n",
       "      <td>1.487627</td>\n",
       "      <td>1.259056</td>\n",
       "      <td>1.346771</td>\n",
       "      <td>1.385682</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.363949</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.363949</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V843</th>\n",
       "      <td>1.358463</td>\n",
       "      <td>1.331599</td>\n",
       "      <td>1.205516</td>\n",
       "      <td>1.205516</td>\n",
       "      <td>1.272773</td>\n",
       "      <td>1.082206</td>\n",
       "      <td>1.240346</td>\n",
       "      <td>0.842120</td>\n",
       "      <td>1.383875</td>\n",
       "      <td>1.082206</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.325776</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>843 rows × 338704 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr18-31634804-31635104  chr1-165460961-165461261  \\\n",
       "V1                   2.112241                  1.988862   \n",
       "V2                   2.184422                  2.079218   \n",
       "V3                   2.381858                  1.865213   \n",
       "V4                   2.296301                  1.952485   \n",
       "V5                   2.021073                  1.682703   \n",
       "...                       ...                       ...   \n",
       "V839                 3.323174                  3.171114   \n",
       "V840                 2.806491                  2.806491   \n",
       "V841                 2.915099                  1.888067   \n",
       "V842                 1.385682                  1.421908   \n",
       "V843                 1.358463                  1.331599   \n",
       "\n",
       "      chr1-75317415-75317715  chr18-46525819-46526119  chr4-3938558-3938858  \\\n",
       "V1                  1.988862                 1.664337              1.758608   \n",
       "V2                  1.957768                 1.533112              1.814123   \n",
       "V3                  1.717455                 1.542694              2.298454   \n",
       "V4                  1.863144                 1.952485              1.364729   \n",
       "V5                  1.763392                 1.903019              1.763392   \n",
       "...                      ...                      ...                   ...   \n",
       "V839                3.171114                 2.585932              2.322799   \n",
       "V840                2.055718                 2.656048              2.656048   \n",
       "V841                1.626292                 2.101953              2.577613   \n",
       "V842                1.021733                 1.209012              1.346771   \n",
       "V843                1.205516                 1.205516              1.272773   \n",
       "\n",
       "      chr6-71440583-71440883  chr7-105399959-105400259  \\\n",
       "V1                  2.112241                  1.758608   \n",
       "V2                  1.889215                  1.957768   \n",
       "V3                  1.542694                  2.298454   \n",
       "V4                  2.107085                  1.763270   \n",
       "V5                  1.836383                  1.682703   \n",
       "...                      ...                       ...   \n",
       "V839                2.585932                  3.323174   \n",
       "V840                2.485490                  1.770699   \n",
       "V841                1.888067                  2.282790   \n",
       "V842                1.385682                  1.487627   \n",
       "V843                1.082206                  1.240346   \n",
       "\n",
       "      chr1-161876704-161877004  chr6-113306525-113306825  \\\n",
       "V1                    1.434083                  1.919221   \n",
       "V2                    1.814123                  1.731114   \n",
       "V3                    1.993207                  1.865213   \n",
       "V4                    1.952485                  1.863144   \n",
       "V5                    1.763392                  1.490239   \n",
       "...                        ...                       ...   \n",
       "V839                  2.585932                  2.000750   \n",
       "V840                  2.055718                  2.806491   \n",
       "V841                  1.626292                  2.439439   \n",
       "V842                  1.259056                  1.346771   \n",
       "V843                  0.842120                  1.383875   \n",
       "\n",
       "      chr1-177796516-177796816  ...  chr10-111689260-111689560  \\\n",
       "V1                    1.919221  ...                        0.0   \n",
       "V2                    1.814123  ...                        0.0   \n",
       "V3                    1.865213  ...                        0.0   \n",
       "V4                    2.174958  ...                        0.0   \n",
       "V5                    1.490239  ...                        0.0   \n",
       "...                        ...  ...                        ...   \n",
       "V839                  2.808408  ...                        0.0   \n",
       "V840                  2.288595  ...                        0.0   \n",
       "V841                  2.282790  ...                        0.0   \n",
       "V842                  1.385682  ...                        0.0   \n",
       "V843                  1.082206  ...                        0.0   \n",
       "\n",
       "      chr10-111388777-111389077  chr10-110841761-110842061  \\\n",
       "V1                     0.000000                        0.0   \n",
       "V2                     0.000000                        0.0   \n",
       "V3                     0.000000                        0.0   \n",
       "V4                     0.000000                        0.0   \n",
       "V5                     0.000000                        0.0   \n",
       "...                         ...                        ...   \n",
       "V839                   0.000000                        0.0   \n",
       "V840                   0.000000                        0.0   \n",
       "V841                   0.000000                        0.0   \n",
       "V842                   0.363949                        0.0   \n",
       "V843                   0.325776                        0.0   \n",
       "\n",
       "      chr10-10884983-10885283  chr10-108335612-108335912  \\\n",
       "V1                        0.0                   0.554779   \n",
       "V2                        0.0                   0.546106   \n",
       "V3                        0.0                   0.000000   \n",
       "V4                        0.0                   0.000000   \n",
       "V5                        0.0                   0.841352   \n",
       "...                       ...                        ...   \n",
       "V839                      0.0                   0.000000   \n",
       "V840                      0.0                   0.000000   \n",
       "V841                      0.0                   0.000000   \n",
       "V842                      0.0                   0.000000   \n",
       "V843                      0.0                   0.000000   \n",
       "\n",
       "      chr10-107887768-107888068  chr10-107038880-107039180  \\\n",
       "V1                          0.0                        0.0   \n",
       "V2                          0.0                        0.0   \n",
       "V3                          0.0                        0.0   \n",
       "V4                          0.0                        0.0   \n",
       "V5                          0.0                        0.0   \n",
       "...                         ...                        ...   \n",
       "V839                        0.0                        0.0   \n",
       "V840                        0.0                        0.0   \n",
       "V841                        0.0                        0.0   \n",
       "V842                        0.0                        0.0   \n",
       "V843                        0.0                        0.0   \n",
       "\n",
       "      chr10-10549625-10549925  chr10-105270865-105271165  \\\n",
       "V1                   0.000000                        0.0   \n",
       "V2                   0.000000                        0.0   \n",
       "V3                   0.000000                        0.0   \n",
       "V4                   0.000000                        0.0   \n",
       "V5                   0.000000                        0.0   \n",
       "...                       ...                        ...   \n",
       "V839                 1.000375                        0.0   \n",
       "V840                 0.000000                        0.0   \n",
       "V841                 0.813146                        0.0   \n",
       "V842                 0.363949                        0.0   \n",
       "V843                 0.000000                        0.0   \n",
       "\n",
       "      chr10-103291333-103291633  \n",
       "V1                          0.0  \n",
       "V2                          0.0  \n",
       "V3                          0.0  \n",
       "V4                          0.0  \n",
       "V5                          0.0  \n",
       "...                         ...  \n",
       "V839                        0.0  \n",
       "V840                        0.0  \n",
       "V841                        0.0  \n",
       "V842                        0.0  \n",
       "V843                        0.0  \n",
       "\n",
       "[843 rows x 338704 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_x = pd.read_csv('atac.aggregate_30cells.csv', index_col=0).transpose()\n",
    "df_x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f0b18be7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0610009B22Rik</th>\n",
       "      <th>0610009L18Rik</th>\n",
       "      <th>0610010F05Rik</th>\n",
       "      <th>0610010K14Rik</th>\n",
       "      <th>0610012D04Rik</th>\n",
       "      <th>0610012G03Rik</th>\n",
       "      <th>0610025J13Rik</th>\n",
       "      <th>0610030E20Rik</th>\n",
       "      <th>0610038B21Rik</th>\n",
       "      <th>0610039K10Rik</th>\n",
       "      <th>...</th>\n",
       "      <th>mt-Co2</th>\n",
       "      <th>mt-Co3</th>\n",
       "      <th>mt-Cytb</th>\n",
       "      <th>mt-Nd1</th>\n",
       "      <th>mt-Nd2</th>\n",
       "      <th>mt-Nd3</th>\n",
       "      <th>mt-Nd4</th>\n",
       "      <th>mt-Nd4l</th>\n",
       "      <th>mt-Nd5</th>\n",
       "      <th>mt-Nd6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>V1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.179400</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.589700</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.629728</td>\n",
       "      <td>2.459005</td>\n",
       "      <td>1.869305</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.369241</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.667545</td>\n",
       "      <td>1.655498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.547446</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.191146</td>\n",
       "      <td>1.849178</td>\n",
       "      <td>1.713501</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.547446</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.063261</td>\n",
       "      <td>1.333363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.581185</td>\n",
       "      <td>1.162371</td>\n",
       "      <td>0.581185</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.581185</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.468833</td>\n",
       "      <td>2.591756</td>\n",
       "      <td>2.270628</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.930656</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.511841</td>\n",
       "      <td>1.842314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.265819</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.003138</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.822417</td>\n",
       "      <td>2.409712</td>\n",
       "      <td>1.776802</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.102480</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.688554</td>\n",
       "      <td>2.006276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V5</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.897976</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.566560</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.566560</td>\n",
       "      <td>0.566560</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.526534</td>\n",
       "      <td>2.266239</td>\n",
       "      <td>2.157094</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.882071</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.448630</td>\n",
       "      <td>1.699679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V839</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.287068</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.039954</td>\n",
       "      <td>3.861204</td>\n",
       "      <td>2.574136</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.574136</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.275547</td>\n",
       "      <td>2.039954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V840</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.502798</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.502798</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.670258</td>\n",
       "      <td>1.670258</td>\n",
       "      <td>1.167460</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.593831</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.208447</td>\n",
       "      <td>1.299713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V841</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.453196</td>\n",
       "      <td>2.113068</td>\n",
       "      <td>2.113068</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.674567</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.787635</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V842</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.932346</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.571665</td>\n",
       "      <td>0.360681</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.558835</td>\n",
       "      <td>1.608430</td>\n",
       "      <td>1.247750</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.143330</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.733919</td>\n",
       "      <td>0.837474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V843</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.445646</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.281171</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.562343</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.215202</td>\n",
       "      <td>1.336938</td>\n",
       "      <td>1.007989</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.843514</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.553067</td>\n",
       "      <td>0.934031</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>843 rows × 19423 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0610009B22Rik  0610009L18Rik  0610010F05Rik  0610010K14Rik  \\\n",
       "V1              0.0       0.000000       1.179400       0.000000   \n",
       "V2              0.0       0.000000       1.547446       0.000000   \n",
       "V3              0.0       0.581185       1.162371       0.581185   \n",
       "V4              0.0       0.000000       1.265819       0.000000   \n",
       "V5              0.0       0.000000       0.897976       0.000000   \n",
       "...             ...            ...            ...            ...   \n",
       "V839            0.0       0.000000       1.287068       0.000000   \n",
       "V840            0.0       0.000000       0.502798       0.000000   \n",
       "V841            0.0       0.000000       0.000000       0.000000   \n",
       "V842            0.0       0.000000       0.932346       0.000000   \n",
       "V843            0.0       0.000000       0.445646       0.000000   \n",
       "\n",
       "      0610012D04Rik  0610012G03Rik  0610025J13Rik  0610030E20Rik  \\\n",
       "V1              0.0       0.000000            0.0       0.589700   \n",
       "V2              0.0       0.000000            0.0       0.000000   \n",
       "V3              0.0       0.000000            0.0       0.000000   \n",
       "V4              0.0       0.000000            0.0       1.003138   \n",
       "V5              0.0       0.566560            0.0       0.566560   \n",
       "...             ...            ...            ...            ...   \n",
       "V839            0.0       0.000000            0.0       0.000000   \n",
       "V840            0.0       0.000000            0.0       0.502798   \n",
       "V841            0.0       0.000000            0.0       0.000000   \n",
       "V842            0.0       0.000000            0.0       0.571665   \n",
       "V843            0.0       0.281171            0.0       0.562343   \n",
       "\n",
       "      0610038B21Rik  0610039K10Rik  ...  mt-Co2  mt-Co3   mt-Cytb    mt-Nd1  \\\n",
       "V1         0.000000            0.0  ...     0.0     0.0  2.629728  2.459005   \n",
       "V2         0.000000            0.0  ...     0.0     0.0  2.191146  1.849178   \n",
       "V3         0.581185            0.0  ...     0.0     0.0  2.468833  2.591756   \n",
       "V4         0.000000            0.0  ...     0.0     0.0  2.822417  2.409712   \n",
       "V5         0.566560            0.0  ...     0.0     0.0  2.526534  2.266239   \n",
       "...             ...            ...  ...     ...     ...       ...       ...   \n",
       "V839       0.000000            0.0  ...     0.0     0.0  2.039954  3.861204   \n",
       "V840       0.000000            0.0  ...     0.0     0.0  1.670258  1.670258   \n",
       "V841       0.000000            0.0  ...     0.0     0.0  2.453196  2.113068   \n",
       "V842       0.360681            0.0  ...     0.0     0.0  1.558835  1.608430   \n",
       "V843       0.000000            0.0  ...     0.0     0.0  1.215202  1.336938   \n",
       "\n",
       "        mt-Nd2  mt-Nd3    mt-Nd4  mt-Nd4l    mt-Nd5    mt-Nd6  \n",
       "V1    1.869305     0.0  1.369241      0.0  2.667545  1.655498  \n",
       "V2    1.713501     0.0  1.547446      0.0  2.063261  1.333363  \n",
       "V3    2.270628     0.0  1.930656      0.0  2.511841  1.842314  \n",
       "V4    1.776802     0.0  2.102480      0.0  2.688554  2.006276  \n",
       "V5    2.157094     0.0  1.882071      0.0  2.448630  1.699679  \n",
       "...        ...     ...       ...      ...       ...       ...  \n",
       "V839  2.574136     0.0  2.574136      0.0  4.275547  2.039954  \n",
       "V840  1.167460     0.0  1.593831      0.0  2.208447  1.299713  \n",
       "V841  2.113068     0.0  1.674567      0.0  3.787635  0.000000  \n",
       "V842  1.247750     0.0  1.143330      0.0  1.733919  0.837474  \n",
       "V843  1.007989     0.0  0.843514      0.0  1.553067  0.934031  \n",
       "\n",
       "[843 rows x 19423 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_y = pd.read_csv('rna.aggregate_30cells.csv', index_col=0).transpose()\n",
    "df_y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c4173184",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr18</td>\n",
       "      <td>31634804</td>\n",
       "      <td>31635104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>165460961</td>\n",
       "      <td>165461261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>75317415</td>\n",
       "      <td>75317715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr18</td>\n",
       "      <td>46525819</td>\n",
       "      <td>46526119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr4</td>\n",
       "      <td>3938558</td>\n",
       "      <td>3938858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338699</th>\n",
       "      <td>chr10</td>\n",
       "      <td>107887768</td>\n",
       "      <td>107888068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338700</th>\n",
       "      <td>chr10</td>\n",
       "      <td>107038880</td>\n",
       "      <td>107039180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338701</th>\n",
       "      <td>chr10</td>\n",
       "      <td>10549625</td>\n",
       "      <td>10549925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338702</th>\n",
       "      <td>chr10</td>\n",
       "      <td>105270865</td>\n",
       "      <td>105271165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338703</th>\n",
       "      <td>chr10</td>\n",
       "      <td>103291333</td>\n",
       "      <td>103291633</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>338704 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom      start        end\n",
       "0       chr18   31634804   31635104\n",
       "1        chr1  165460961  165461261\n",
       "2        chr1   75317415   75317715\n",
       "3       chr18   46525819   46526119\n",
       "4        chr4    3938558    3938858\n",
       "...       ...        ...        ...\n",
       "338699  chr10  107887768  107888068\n",
       "338700  chr10  107038880  107039180\n",
       "338701  chr10   10549625   10549925\n",
       "338702  chr10  105270865  105271165\n",
       "338703  chr10  103291333  103291633\n",
       "\n",
       "[338704 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_peaks = pd.DataFrame(df_x.columns)[0].str.split('-',expand=True) \n",
    "df_peaks = df_peaks.rename(columns={0: \"chrom\", 1: \"start\", 2: \"end\"})\n",
    "df_peaks[\"start\"] = pd.to_numeric(df_peaks[\"start\"])\n",
    "df_peaks[\"end\"] = pd.to_numeric(df_peaks[\"end\"])\n",
    "df_peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8f5f964e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_peaks.to_csv('/data1/xixi/scRegulate/multiomic_data/10x_PBMC/peaks_all.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "81c0be67",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene stable ID version</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Source of gene name</th>\n",
       "      <th>Transcription start site (TSS)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000070103</td>\n",
       "      <td>ENSMUSG00000070103.2</td>\n",
       "      <td>1</td>\n",
       "      <td>158505625</td>\n",
       "      <td>158505733</td>\n",
       "      <td>1</td>\n",
       "      <td>Mir488</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>158505625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000065567</td>\n",
       "      <td>ENSMUSG00000065567.1</td>\n",
       "      <td>1</td>\n",
       "      <td>23291701</td>\n",
       "      <td>23291784</td>\n",
       "      <td>1</td>\n",
       "      <td>Mir30c-2</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>23291701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000094946</td>\n",
       "      <td>ENSMUSG00000094946.1</td>\n",
       "      <td>1</td>\n",
       "      <td>83795912</td>\n",
       "      <td>83796032</td>\n",
       "      <td>1</td>\n",
       "      <td>Gm25754</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>83795912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000093155</td>\n",
       "      <td>ENSMUSG00000093155.1</td>\n",
       "      <td>1</td>\n",
       "      <td>74586896</td>\n",
       "      <td>74586972</td>\n",
       "      <td>-1</td>\n",
       "      <td>Gm25035</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>74586972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000065458</td>\n",
       "      <td>ENSMUSG00000065458.1</td>\n",
       "      <td>1</td>\n",
       "      <td>137966639</td>\n",
       "      <td>137966718</td>\n",
       "      <td>1</td>\n",
       "      <td>Mir181b-1</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>137966639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103991</th>\n",
       "      <td>ENSMUSG00000099633</td>\n",
       "      <td>ENSMUSG00000099633.1</td>\n",
       "      <td>Y</td>\n",
       "      <td>84907473</td>\n",
       "      <td>84910700</td>\n",
       "      <td>1</td>\n",
       "      <td>Gm29071</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>84907473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103993</th>\n",
       "      <td>ENSMUSG00000100388</td>\n",
       "      <td>ENSMUSG00000100388.1</td>\n",
       "      <td>Y</td>\n",
       "      <td>50770044</td>\n",
       "      <td>50773283</td>\n",
       "      <td>1</td>\n",
       "      <td>Gm29116</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>50770044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103995</th>\n",
       "      <td>ENSMUSG00000091987</td>\n",
       "      <td>ENSMUSG00000091987.8</td>\n",
       "      <td>Y</td>\n",
       "      <td>2900989</td>\n",
       "      <td>2912206</td>\n",
       "      <td>1</td>\n",
       "      <td>Gm10352</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>2900989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103997</th>\n",
       "      <td>ENSMUSG00000101667</td>\n",
       "      <td>ENSMUSG00000101667.1</td>\n",
       "      <td>Y</td>\n",
       "      <td>2932582</td>\n",
       "      <td>2939416</td>\n",
       "      <td>1</td>\n",
       "      <td>Gm29289</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>2932582</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103998</th>\n",
       "      <td>ENSMUSG00000101782</td>\n",
       "      <td>ENSMUSG00000101782.1</td>\n",
       "      <td>Y</td>\n",
       "      <td>57125046</td>\n",
       "      <td>57128288</td>\n",
       "      <td>1</td>\n",
       "      <td>Gm28964</td>\n",
       "      <td>MGI Symbol</td>\n",
       "      <td>57125046</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29642 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Gene stable ID Gene stable ID version Chromosome/scaffold name  \\\n",
       "0       ENSMUSG00000070103   ENSMUSG00000070103.2                        1   \n",
       "1       ENSMUSG00000065567   ENSMUSG00000065567.1                        1   \n",
       "2       ENSMUSG00000094946   ENSMUSG00000094946.1                        1   \n",
       "3       ENSMUSG00000093155   ENSMUSG00000093155.1                        1   \n",
       "4       ENSMUSG00000065458   ENSMUSG00000065458.1                        1   \n",
       "...                    ...                    ...                      ...   \n",
       "103991  ENSMUSG00000099633   ENSMUSG00000099633.1                        Y   \n",
       "103993  ENSMUSG00000100388   ENSMUSG00000100388.1                        Y   \n",
       "103995  ENSMUSG00000091987   ENSMUSG00000091987.8                        Y   \n",
       "103997  ENSMUSG00000101667   ENSMUSG00000101667.1                        Y   \n",
       "103998  ENSMUSG00000101782   ENSMUSG00000101782.1                        Y   \n",
       "\n",
       "        Gene start (bp)  Gene end (bp)  Strand  Gene name Source of gene name  \\\n",
       "0             158505625      158505733       1     Mir488          MGI Symbol   \n",
       "1              23291701       23291784       1   Mir30c-2          MGI Symbol   \n",
       "2              83795912       83796032       1    Gm25754          MGI Symbol   \n",
       "3              74586896       74586972      -1    Gm25035          MGI Symbol   \n",
       "4             137966639      137966718       1  Mir181b-1          MGI Symbol   \n",
       "...                 ...            ...     ...        ...                 ...   \n",
       "103991         84907473       84910700       1    Gm29071          MGI Symbol   \n",
       "103993         50770044       50773283       1    Gm29116          MGI Symbol   \n",
       "103995          2900989        2912206       1    Gm10352          MGI Symbol   \n",
       "103997          2932582        2939416       1    Gm29289          MGI Symbol   \n",
       "103998         57125046       57128288       1    Gm28964          MGI Symbol   \n",
       "\n",
       "        Transcription start site (TSS)  \n",
       "0                            158505625  \n",
       "1                             23291701  \n",
       "2                             83795912  \n",
       "3                             74586972  \n",
       "4                            137966639  \n",
       "...                                ...  \n",
       "103991                        84907473  \n",
       "103993                        50770044  \n",
       "103995                         2900989  \n",
       "103997                         2932582  \n",
       "103998                        57125046  \n",
       "\n",
       "[29642 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geneanno = pd.read_csv('../../ref_genome/mm10_geneanno.txt', sep='\\t')\n",
    "geneanno = geneanno.drop_duplicates(subset=['Gene name'])\n",
    "geneanno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a41e75d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MA0002.2.Runx1.npy',\n",
       " 'MA0006.1.Ahr::Arnt.npy',\n",
       " 'MA0004.1.Arnt.npy',\n",
       " 'MA0007.3.Ar.npy',\n",
       " 'MA0009.1.Tbxt.npy']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "motif_files = os.listdir('../../ref_genome/JASPAR_motifs_pfm_mouse/pfm.np')\n",
    "motif_files[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3aea3ca3",
   "metadata": {},
   "outputs": [],
   "source": [
    "tfs_kept = []\n",
    "tf_by_region_mat = []\n",
    "for i in list(motif_files):\n",
    "    tf = i.split('.')[-2].capitalize()\n",
    "    if tf in df_y.columns:\n",
    "        if tf in tfs_kept:\n",
    "            continue\n",
    "            \n",
    "        tfs_kept.append(tf)\n",
    "tfs_kept = sorted(tfs_kept)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "90321fdf",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id1</th>\n",
       "      <th>name</th>\n",
       "      <th>namespace</th>\n",
       "      <th>def</th>\n",
       "      <th>relation</th>\n",
       "      <th>id2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GO:0031069</td>\n",
       "      <td>hair follicle morphogenesis</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>The process in which the anatomical structures...</td>\n",
       "      <td>part_of</td>\n",
       "      <td>GO:0001942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GO:0048820</td>\n",
       "      <td>hair follicle maturation</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>A developmental process, independent of morpho...</td>\n",
       "      <td>part_of</td>\n",
       "      <td>GO:0001942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GO:0051797</td>\n",
       "      <td>regulation of hair follicle development</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>Any process that modulates the frequency, rate...</td>\n",
       "      <td>regulates</td>\n",
       "      <td>GO:0001942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GO:0051798</td>\n",
       "      <td>positive regulation of hair follicle development</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>Any process that activates or increases the fr...</td>\n",
       "      <td>positively_regulates</td>\n",
       "      <td>GO:0001942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GO:0051799</td>\n",
       "      <td>negative regulation of hair follicle development</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>Any process that stops, prevents, or reduces t...</td>\n",
       "      <td>negatively_regulates</td>\n",
       "      <td>GO:0001942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GO:0060789</td>\n",
       "      <td>hair follicle placode formation</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>The developmental process in which a hair plac...</td>\n",
       "      <td>part_of</td>\n",
       "      <td>GO:0001942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GO:0042637</td>\n",
       "      <td>catagen</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>The regression phase of the hair cycle during ...</td>\n",
       "      <td>part_of</td>\n",
       "      <td>GO:0048820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GO:0042640</td>\n",
       "      <td>anagen</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>The growth phase of the hair cycle. Lasts, for...</td>\n",
       "      <td>part_of</td>\n",
       "      <td>GO:0048820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GO:0051798</td>\n",
       "      <td>positive regulation of hair follicle development</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>Any process that activates or increases the fr...</td>\n",
       "      <td>is_a</td>\n",
       "      <td>GO:0051797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GO:0051799</td>\n",
       "      <td>negative regulation of hair follicle development</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>Any process that stops, prevents, or reduces t...</td>\n",
       "      <td>is_a</td>\n",
       "      <td>GO:0051797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GO:0051795</td>\n",
       "      <td>positive regulation of timing of catagen</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>Any process that activates or increases the fr...</td>\n",
       "      <td>positively_regulates</td>\n",
       "      <td>GO:0042637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GO:0051885</td>\n",
       "      <td>positive regulation of timing of anagen</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>Any process that activates or increases the fr...</td>\n",
       "      <td>positively_regulates</td>\n",
       "      <td>GO:0042640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GO:0051886</td>\n",
       "      <td>negative regulation of timing of anagen</td>\n",
       "      <td>biological_process</td>\n",
       "      <td>Any process that stops, prevents, or reduces t...</td>\n",
       "      <td>negatively_regulates</td>\n",
       "      <td>GO:0042640</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           id1                                              name  \\\n",
       "0   GO:0031069                       hair follicle morphogenesis   \n",
       "1   GO:0048820                          hair follicle maturation   \n",
       "2   GO:0051797           regulation of hair follicle development   \n",
       "3   GO:0051798  positive regulation of hair follicle development   \n",
       "4   GO:0051799  negative regulation of hair follicle development   \n",
       "5   GO:0060789                   hair follicle placode formation   \n",
       "6   GO:0042637                                           catagen   \n",
       "7   GO:0042640                                            anagen   \n",
       "8   GO:0051798  positive regulation of hair follicle development   \n",
       "9   GO:0051799  negative regulation of hair follicle development   \n",
       "10  GO:0051795          positive regulation of timing of catagen   \n",
       "11  GO:0051885           positive regulation of timing of anagen   \n",
       "12  GO:0051886           negative regulation of timing of anagen   \n",
       "\n",
       "             namespace                                                def  \\\n",
       "0   biological_process  The process in which the anatomical structures...   \n",
       "1   biological_process  A developmental process, independent of morpho...   \n",
       "2   biological_process  Any process that modulates the frequency, rate...   \n",
       "3   biological_process  Any process that activates or increases the fr...   \n",
       "4   biological_process  Any process that stops, prevents, or reduces t...   \n",
       "5   biological_process  The developmental process in which a hair plac...   \n",
       "6   biological_process  The regression phase of the hair cycle during ...   \n",
       "7   biological_process  The growth phase of the hair cycle. Lasts, for...   \n",
       "8   biological_process  Any process that activates or increases the fr...   \n",
       "9   biological_process  Any process that stops, prevents, or reduces t...   \n",
       "10  biological_process  Any process that activates or increases the fr...   \n",
       "11  biological_process  Any process that activates or increases the fr...   \n",
       "12  biological_process  Any process that stops, prevents, or reduces t...   \n",
       "\n",
       "                relation         id2  \n",
       "0                part_of  GO:0001942  \n",
       "1                part_of  GO:0001942  \n",
       "2              regulates  GO:0001942  \n",
       "3   positively_regulates  GO:0001942  \n",
       "4   negatively_regulates  GO:0001942  \n",
       "5                part_of  GO:0001942  \n",
       "6                part_of  GO:0048820  \n",
       "7                part_of  GO:0048820  \n",
       "8                   is_a  GO:0051797  \n",
       "9                   is_a  GO:0051797  \n",
       "10  positively_regulates  GO:0042637  \n",
       "11  positively_regulates  GO:0042640  \n",
       "12  negatively_regulates  GO:0042640  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go = pd.read_csv('./go/go_follicle.txt', sep='\\t')\n",
    "go"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f53654a7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X0</th>\n",
       "      <th>X1</th>\n",
       "      <th>X2</th>\n",
       "      <th>X3</th>\n",
       "      <th>X4</th>\n",
       "      <th>X5</th>\n",
       "      <th>X6</th>\n",
       "      <th>X7</th>\n",
       "      <th>X8</th>\n",
       "      <th>X9</th>\n",
       "      <th>X10</th>\n",
       "      <th>X11</th>\n",
       "      <th>X12</th>\n",
       "      <th>X13</th>\n",
       "      <th>X14</th>\n",
       "      <th>X15</th>\n",
       "      <th>X16</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:108359</td>\n",
       "      <td>Alx4</td>\n",
       "      <td>involved_in</td>\n",
       "      <td>GO:0001942</td>\n",
       "      <td>MGI:MGI:4834177|GO_REF:0000096</td>\n",
       "      <td>ISO</td>\n",
       "      <td>UniProtKB:Q9H161</td>\n",
       "      <td>P</td>\n",
       "      <td>aristaless-like homeobox 4</td>\n",
       "      <td>Aristaless-like 4</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>taxon:10090</td>\n",
       "      <td>20100901</td>\n",
       "      <td>MGI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:3513977</td>\n",
       "      <td>Apcdd1</td>\n",
       "      <td>involved_in</td>\n",
       "      <td>GO:0001942</td>\n",
       "      <td>MGI:MGI:4834177|GO_REF:0000096</td>\n",
       "      <td>ISO</td>\n",
       "      <td>UniProtKB:Q8J025</td>\n",
       "      <td>P</td>\n",
       "      <td>adenomatosis polyposis coli down-regulated 1</td>\n",
       "      <td>Drapc1|EIG180</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>taxon:10090</td>\n",
       "      <td>20100423</td>\n",
       "      <td>MGI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:88138</td>\n",
       "      <td>Bcl2</td>\n",
       "      <td>involved_in</td>\n",
       "      <td>GO:0031069</td>\n",
       "      <td>MGI:MGI:63353|PMID:8402909</td>\n",
       "      <td>IMP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P</td>\n",
       "      <td>B cell leukemia/lymphoma 2</td>\n",
       "      <td>Bcl-2|C430015F12Rik|D830018M01Rik</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>taxon:10090</td>\n",
       "      <td>20070305</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:2443583</td>\n",
       "      <td>Fermt1</td>\n",
       "      <td>involved_in</td>\n",
       "      <td>GO:0051886</td>\n",
       "      <td>MGI:MGI:5571208|PMID:24681597</td>\n",
       "      <td>IMP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P</td>\n",
       "      <td>fermitin family member 1</td>\n",
       "      <td>5830467P10Rik|Kindlin-1</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>taxon:10090</td>\n",
       "      <td>20161121</td>\n",
       "      <td>CAFA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:102949</td>\n",
       "      <td>Foxn1</td>\n",
       "      <td>involved_in</td>\n",
       "      <td>GO:0051798</td>\n",
       "      <td>MGI:MGI:5437206|PMID:21109991</td>\n",
       "      <td>IMP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P</td>\n",
       "      <td>forkhead box N1</td>\n",
       "      <td>D11Bhm185e|Hfh11|whn</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>taxon:10090</td>\n",
       "      <td>20150111</td>\n",
       "      <td>UniProt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:99400</td>\n",
       "      <td>Atp7a</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>GO:0031069</td>\n",
       "      <td>PMID:2473662</td>\n",
       "      <td>IMP</td>\n",
       "      <td>MGI:MGI:1856466</td>\n",
       "      <td>P</td>\n",
       "      <td>ATPase, Cu++ transporting, alpha polypeptide</td>\n",
       "      <td>Menkes protein|MNK|br</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>taxon:10090</td>\n",
       "      <td>20061107</td>\n",
       "      <td>MGI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:99560</td>\n",
       "      <td>Hoxc13</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>GO:0001942</td>\n",
       "      <td>PMID:21191399</td>\n",
       "      <td>IMP</td>\n",
       "      <td>MGI:MGI:1926422</td>\n",
       "      <td>P</td>\n",
       "      <td>homeobox C13</td>\n",
       "      <td>N</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>taxon:10090</td>\n",
       "      <td>20121227</td>\n",
       "      <td>MGI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:99560</td>\n",
       "      <td>Hoxc13</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>GO:0001942</td>\n",
       "      <td>PMID:9420327</td>\n",
       "      <td>IMP</td>\n",
       "      <td>MGI:MGI:1926422</td>\n",
       "      <td>P</td>\n",
       "      <td>homeobox C13</td>\n",
       "      <td>N</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>taxon:10090</td>\n",
       "      <td>20080322</td>\n",
       "      <td>MGI</td>\n",
       "      <td>results_in_development_of(EMAPA:26747)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:99560</td>\n",
       "      <td>Hoxc13</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>GO:0001942</td>\n",
       "      <td>PMID:9420327</td>\n",
       "      <td>IMP</td>\n",
       "      <td>MGI:MGI:1926422</td>\n",
       "      <td>P</td>\n",
       "      <td>homeobox C13</td>\n",
       "      <td>N</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>taxon:10090</td>\n",
       "      <td>20080322</td>\n",
       "      <td>MGI</td>\n",
       "      <td>results_in_development_of(EMAPA:36498)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>MGI</td>\n",
       "      <td>MGI:99852</td>\n",
       "      <td>Runx1</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>GO:0031069</td>\n",
       "      <td>PMID:17011173</td>\n",
       "      <td>IMP</td>\n",
       "      <td>MGI:MGI:3043614</td>\n",
       "      <td>P</td>\n",
       "      <td>runt related transcription factor 1</td>\n",
       "      <td>runt domain, alpha subunit 2|Pebp2a2|AML1|Cbfa2</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>taxon:10090</td>\n",
       "      <td>20111214</td>\n",
       "      <td>MGI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>181 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      X0           X1      X2                          X3          X4  \\\n",
       "0    MGI   MGI:108359    Alx4                 involved_in  GO:0001942   \n",
       "1    MGI  MGI:3513977  Apcdd1                 involved_in  GO:0001942   \n",
       "2    MGI    MGI:88138    Bcl2                 involved_in  GO:0031069   \n",
       "3    MGI  MGI:2443583  Fermt1                 involved_in  GO:0051886   \n",
       "4    MGI   MGI:102949   Foxn1                 involved_in  GO:0051798   \n",
       "..   ...          ...     ...                         ...         ...   \n",
       "176  MGI    MGI:99400   Atp7a  acts_upstream_of_or_within  GO:0031069   \n",
       "177  MGI    MGI:99560  Hoxc13  acts_upstream_of_or_within  GO:0001942   \n",
       "178  MGI    MGI:99560  Hoxc13  acts_upstream_of_or_within  GO:0001942   \n",
       "179  MGI    MGI:99560  Hoxc13  acts_upstream_of_or_within  GO:0001942   \n",
       "180  MGI    MGI:99852   Runx1  acts_upstream_of_or_within  GO:0031069   \n",
       "\n",
       "                                 X5   X6                X7 X8  \\\n",
       "0    MGI:MGI:4834177|GO_REF:0000096  ISO  UniProtKB:Q9H161  P   \n",
       "1    MGI:MGI:4834177|GO_REF:0000096  ISO  UniProtKB:Q8J025  P   \n",
       "2        MGI:MGI:63353|PMID:8402909  IMP               NaN  P   \n",
       "3     MGI:MGI:5571208|PMID:24681597  IMP               NaN  P   \n",
       "4     MGI:MGI:5437206|PMID:21109991  IMP               NaN  P   \n",
       "..                              ...  ...               ... ..   \n",
       "176                    PMID:2473662  IMP   MGI:MGI:1856466  P   \n",
       "177                   PMID:21191399  IMP   MGI:MGI:1926422  P   \n",
       "178                    PMID:9420327  IMP   MGI:MGI:1926422  P   \n",
       "179                    PMID:9420327  IMP   MGI:MGI:1926422  P   \n",
       "180                   PMID:17011173  IMP   MGI:MGI:3043614  P   \n",
       "\n",
       "                                               X9  \\\n",
       "0                      aristaless-like homeobox 4   \n",
       "1    adenomatosis polyposis coli down-regulated 1   \n",
       "2                      B cell leukemia/lymphoma 2   \n",
       "3                        fermitin family member 1   \n",
       "4                                 forkhead box N1   \n",
       "..                                            ...   \n",
       "176  ATPase, Cu++ transporting, alpha polypeptide   \n",
       "177                                  homeobox C13   \n",
       "178                                  homeobox C13   \n",
       "179                                  homeobox C13   \n",
       "180           runt related transcription factor 1   \n",
       "\n",
       "                                                 X10                  X11  \\\n",
       "0                                  Aristaless-like 4  protein_coding_gene   \n",
       "1                                      Drapc1|EIG180  protein_coding_gene   \n",
       "2                  Bcl-2|C430015F12Rik|D830018M01Rik  protein_coding_gene   \n",
       "3                            5830467P10Rik|Kindlin-1  protein_coding_gene   \n",
       "4                               D11Bhm185e|Hfh11|whn  protein_coding_gene   \n",
       "..                                               ...                  ...   \n",
       "176                            Menkes protein|MNK|br  protein_coding_gene   \n",
       "177                                                N  protein_coding_gene   \n",
       "178                                                N  protein_coding_gene   \n",
       "179                                                N  protein_coding_gene   \n",
       "180  runt domain, alpha subunit 2|Pebp2a2|AML1|Cbfa2  protein_coding_gene   \n",
       "\n",
       "             X12       X13      X14                                     X15  \\\n",
       "0    taxon:10090  20100901      MGI                                     NaN   \n",
       "1    taxon:10090  20100423      MGI                                     NaN   \n",
       "2    taxon:10090  20070305  UniProt                                     NaN   \n",
       "3    taxon:10090  20161121     CAFA                                     NaN   \n",
       "4    taxon:10090  20150111  UniProt                                     NaN   \n",
       "..           ...       ...      ...                                     ...   \n",
       "176  taxon:10090  20061107      MGI                                     NaN   \n",
       "177  taxon:10090  20121227      MGI                                     NaN   \n",
       "178  taxon:10090  20080322      MGI  results_in_development_of(EMAPA:26747)   \n",
       "179  taxon:10090  20080322      MGI  results_in_development_of(EMAPA:36498)   \n",
       "180  taxon:10090  20111214      MGI                                     NaN   \n",
       "\n",
       "     X16  \n",
       "0    NaN  \n",
       "1    NaN  \n",
       "2    NaN  \n",
       "3    NaN  \n",
       "4    NaN  \n",
       "..   ...  \n",
       "176  NaN  \n",
       "177  NaN  \n",
       "178  NaN  \n",
       "179  NaN  \n",
       "180  NaN  \n",
       "\n",
       "[181 rows x 17 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "goa = pd.read_csv('./go/goa_follicle.txt', sep='\\t')\n",
    "goa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "fd9c12b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "77"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files = os.listdir('/nfs/public/xixi/scRegulate/SHAREseq/nn.best.feature6.learnW_go')\n",
    "markers_filtered = []\n",
    "for file in files:\n",
    "    marker = file.split('.')[0]\n",
    "    if marker not in markers_filtered:\n",
    "        markers_filtered.append(marker)\n",
    "len(markers_filtered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2a3d204a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "113\n",
      "107\n",
      "77\n"
     ]
    }
   ],
   "source": [
    "genes = goa['X2'].drop_duplicates().to_list()\n",
    "print(len(genes))\n",
    "genes_filtered = [i for i in genes if i not in tfs_kept]\n",
    "print(len(genes_filtered))\n",
    "genes_filtered = [i for i in genes_filtered if i in markers_filtered]\n",
    "print(len(genes_filtered))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cae9f3ff",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "goa_filtered = goa.loc[goa['X2'].isin(genes_filtered),]\n",
    "gos = goa_filtered['X4'].drop_duplicates().to_list()\n",
    "len(gos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7830fdb7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_filtered = go.loc[go['id1'].isin(gos) & go['id2'].isin(gos),]\n",
    "len(pd.concat([go_filtered['id1'].drop_duplicates(), go_filtered['id2'].drop_duplicates()]).drop_duplicates())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6294fcc0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "torch.Size([12, 77])"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gos = goa_filtered['X4'].drop_duplicates().to_list()\n",
    "mask = torch.zeros(len(genes_filtered), len(gos))\n",
    "for i in range(len(goa_filtered['X2'])):\n",
    "    if goa_filtered['X2'].iloc[i] in genes_filtered:\n",
    "        mask[genes_filtered.index(goa_filtered['X2'].iloc[i]), gos.index(goa_filtered['X4'].iloc[i])] = 1\n",
    "mask = mask.t()\n",
    "mask.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5ea41e5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],\n",
       "       [1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],\n",
       "       [0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],\n",
       "       [1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],\n",
       "       [1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1.],\n",
       "       [1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],\n",
       "       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0.],\n",
       "       [0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],\n",
       "       [0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.],\n",
       "       [1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1.],\n",
       "       [0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.],\n",
       "       [1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "W_true = np.zeros((len(gos), len(gos)))\n",
    "for i in range(len(go_filtered)):\n",
    "    W_true[gos.index(go_filtered['id1'].iloc[i]), gos.index(go_filtered['id2'].iloc[i])]=1\n",
    "W_true"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ce285407",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(843, 165, 11058)\n",
      "(843, 77)\n",
      "[142 177  60 140 104 156  78 320 256  98 108 205  93 105  73 198 233 209\n",
      " 158 176 272 187 249 357 167 120 152 119 104 243  39  15 257  32 124   7\n",
      " 189 308 115  89 138  54  50 157  74 147 145  74 103 136 110 147  50 131\n",
      " 134 174 173 218 120 146 100  27 119 129 195 153 116  84 140 154 186 132\n",
      " 205  88 182  71 262]\n"
     ]
    }
   ],
   "source": [
    "h5f = h5py.File('./predict_lineage_pseudotime/data_follicle2.h5', 'r')\n",
    "X = h5f['X'][:]\n",
    "expr = h5f['expr'][:]\n",
    "num_peaks = h5f['num_peaks'][:]\n",
    "peaks_all = pd.DataFrame(np.array([item.decode('utf-8') for j in h5f['peaks'][:] for item in j]).reshape(-1, 3))\n",
    "cell_by_tf = h5f['cell_by_tf'][:]\n",
    "cell_by_peak = h5f['cell_by_peak'][:]\n",
    "W = h5f['W'][:]\n",
    "h5f.close()\n",
    "print(X.shape)\n",
    "print(expr.shape)\n",
    "print(num_peaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "fae3f39e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "78"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cut = [0]\n",
    "s = 0\n",
    "for i in num_peaks:\n",
    "    s = s+i\n",
    "    cut.append(s)\n",
    "len(cut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "316565e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0: 0\n",
      "1: 1\n",
      "2: 2\n",
      "3: 3\n",
      "4: 4\n",
      "5: 5\n",
      "(843, 6)\n",
      "(843, 7)\n"
     ]
    }
   ],
   "source": [
    "y = pd.read_csv('skin.aggregate.cellid&cluster&pseudotime_30cells.csv', index_col=0)\n",
    "time = np.array(y['aggr_pseudotime'])\n",
    "y_std = (time - time.min()) / (time.max() - time.min())\n",
    "y_pseudo = y_std * (1 - (0)) + (0)\n",
    "y_pseudo = y_pseudo.reshape((len(y), 1))\n",
    "\n",
    "y = np.array(y['celltype'])\n",
    "y_new = y.reshape((len(y), 1))\n",
    "enc = OneHotEncoder(handle_unknown='ignore')\n",
    "y_oht = enc.fit_transform(y_new).toarray()\n",
    "for i, category in enumerate(enc.categories_[0]):\n",
    "    print(f\"{category}: {i}\")\n",
    "print(y_oht.shape)\n",
    "y_oht = np.concatenate([y_oht, y_pseudo], axis=1)\n",
    "print(y_oht.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6ba2a6e6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(590, 165, 11058)\n",
      "(590, 77)\n",
      "(590, 7)\n",
      "(127, 165, 11058)\n",
      "(126, 165, 11058)\n"
     ]
    }
   ],
   "source": [
    "X_train, X_test, expr_train, expr_test, y_oht_train, y_oht_test, y_train, y_test = train_test_split(X, expr, y_oht, y, test_size=0.3,\n",
    "                                                                                                    random_state=2023, stratify=y)\n",
    "X_test, X_val, expr_test, expr_val, y_oht_test, y_oht_val, y_test, y_val = train_test_split(X_test, expr_test, y_oht_test, y_test, \n",
    "                                                                                                test_size=0.5, random_state=2023, \n",
    "                                                                                            stratify=y_test)\n",
    "print(X_train.shape)\n",
    "print(expr_train.shape)\n",
    "print(y_oht_train.shape)\n",
    "print(X_val.shape)\n",
    "print(X_test.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0191ed86",
   "metadata": {},
   "outputs": [],
   "source": [
    "indices_train, indices_test, y_label_train, y_label_test= train_test_split(np.arange(X.shape[0]), y, \n",
    "                                                                           test_size=0.3, random_state=2023, stratify=y)\n",
    "indices_test, indices_val, y_label_test, y_label_val= train_test_split(indices_test, y_label_test, \n",
    "                                                                           test_size=0.5, random_state=2023, stratify=y_label_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8eb8590b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "torch.Size([590, 165, 11058])\n",
      "torch.Size([590, 77])\n",
      "torch.Size([590, 7])\n"
     ]
    }
   ],
   "source": [
    "X_train = torch.from_numpy(X_train).float()\n",
    "X_val = torch.from_numpy(X_val).float()\n",
    "X_test = torch.from_numpy(X_test).float()\n",
    "expr_train = torch.from_numpy(expr_train).float()\n",
    "expr_val = torch.from_numpy(expr_val).float()\n",
    "expr_test = torch.from_numpy(expr_test).float()\n",
    "y_train = torch.from_numpy(y_oht_train).float()\n",
    "y_val = torch.from_numpy(y_oht_val).float()\n",
    "y_test = torch.from_numpy(y_oht_test).float()\n",
    "print(X_train.shape)\n",
    "print(expr_train.shape)\n",
    "print(y_train.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d280b026",
   "metadata": {},
   "source": [
    "# Def"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "9540fc9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch_geometric.nn import GATConv, GATv2Conv, GCNConv\n",
    "from torch_geometric.nn.conv import MessagePassing\n",
    "from torch.nn import Parameter\n",
    "from torch_geometric.nn.inits import glorot, zeros\n",
    "from torch_geometric.utils import softmax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "dee4f0a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "from torch.nn import Parameter, functional as F\n",
    "from torch_geometric.nn import MessagePassing\n",
    "from torch_geometric.nn.inits import glorot, zeros\n",
    "from torch_geometric.utils import remove_self_loops, add_self_loops, softmax\n",
    "from torch_geometric.utils.num_nodes import maybe_num_nodes\n",
    "#from torch_scatter import scatter_max, scatter_add\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b2e4a109",
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_x = []\n",
    "edge_y = []\n",
    "for i in range(len(go_filtered)):\n",
    "    edge_x.append(gos.index(go_filtered['id1'].iloc[i]))\n",
    "    edge_y.append(gos.index(go_filtered['id2'].iloc[i]))\n",
    "edge_x = np.array(edge_x)\n",
    "edge_y = np.array(edge_y)\n",
    "edge = np.vstack([edge_x, edge_y])\n",
    "edge = torch.from_numpy(edge)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "4f01f1c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "class TensorDataset(Dataset[Tuple[Tensor, ...]]):\n",
    "    r\"\"\"Dataset wrapping tensors.\n",
    "\n",
    "    Each sample will be retrieved by indexing tensors along the first dimension.\n",
    "\n",
    "    Args:\n",
    "        *tensors (Tensor): tensors that have the same size of the first dimension.\n",
    "    \"\"\"\n",
    "    tensors: Tuple[Tensor, ...]\n",
    "\n",
    "    def __init__(self, *tensors: Tensor) -> None:\n",
    "        assert all(tensors[0].size(0) == tensor.size(0) for tensor in tensors), \"Size mismatch between tensors\"\n",
    "        self.tensors = tensors\n",
    "\n",
    "    def __getitem__(self, index):\n",
    "        return tuple(tensor[index] for tensor in self.tensors)\n",
    "\n",
    "    def __len__(self):\n",
    "        return self.tensors[0].size(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "718594eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "class subNet(nn.Module):\n",
    "    def __init__(self, num_peak, num_tf):\n",
    "        super(subNet, self).__init__()\n",
    "        self.num_peak = num_peak\n",
    "        self.num_tf = num_tf\n",
    "\n",
    "        self.fc1 = nn.Linear(self.num_peak * self.num_tf, 1)\n",
    "        self.fc1_activate = nn.LeakyReLU()\n",
    "        self.abs = nn.ReLU()\n",
    "        \n",
    "    def forward(self, x):\n",
    "        x = x.reshape(x.shape[0], -1)\n",
    "        x = self.fc1(x)\n",
    "        return x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "52446c99",
   "metadata": {},
   "outputs": [],
   "source": [
    "class selfGATConv(MessagePassing):\n",
    "    def __init__(self, in_channels, out_channels, heads=2, concat=False,\n",
    "                 negative_slope=0.2, dropout=0, bias=True, **kwargs):\n",
    "        super(selfGATConv, self).__init__(aggr='mean', **kwargs)\n",
    "\n",
    "        self.in_channels = in_channels\n",
    "        self.out_channels = out_channels\n",
    "        self.heads = heads\n",
    "        self.concat = concat\n",
    "        self.negative_slope = negative_slope\n",
    "        self.dropout = dropout\n",
    "        self.node_dim = 0\n",
    "\n",
    "        self.weight = Parameter(\n",
    "            torch.Tensor(in_channels, heads * out_channels))\n",
    "        self.att = Parameter(torch.Tensor(1, 1, heads, 2 * out_channels))\n",
    "\n",
    "        if bias and concat:\n",
    "            self.bias = Parameter(torch.Tensor(heads * out_channels))\n",
    "        elif bias and not concat:\n",
    "            self.bias = Parameter(torch.Tensor(out_channels))\n",
    "        else:\n",
    "            self.register_parameter('bias', None)\n",
    "\n",
    "        self.reset_parameters()\n",
    "\n",
    "    def reset_parameters(self):\n",
    "        glorot(self.weight)\n",
    "        glorot(self.att)\n",
    "        zeros(self.bias)\n",
    "\n",
    "    def forward(self, x, edge_index, size=None):\n",
    "        \"\"\"\"\"\"\n",
    "        # if size is None and torch.is_tensor(x):\n",
    "        #     edge_index, _ = remove_self_loops(edge_index)\n",
    "        #     edge_index, _ = add_self_loops(edge_index, num_nodes=x.size(0))\n",
    "\n",
    "        if torch.is_tensor(x):\n",
    "            x = torch.matmul(x, self.weight)\n",
    "        else:\n",
    "            x = (None if x[0] is None else torch.matmul(x[0], self.weight),\n",
    "                 None if x[1] is None else torch.matmul(x[1], self.weight))\n",
    "\n",
    "        return self.propagate(edge_index, size=size, x=x)\n",
    "\n",
    "    def message(self, edge_index_i, ptr, x_i, x_j, size_i):\n",
    "        # Compute attention coefficients.\n",
    "        x_j = x_j.view(edge_index_i.shape[0], -1, self.heads, self.out_channels)\n",
    "        if x_i is None:\n",
    "            alpha = (x_j * self.att[:, :, :, self.out_channels:]).sum(dim=-1)\n",
    "        else:\n",
    "            x_i = x_i.view(edge_index_i.shape[0], -1, self.heads, self.out_channels)\n",
    "            alpha = (torch.cat([x_i, x_j], dim=-1) * self.att).sum(dim=-1)\n",
    "            #print((torch.cat([x_i, x_j], dim=-1) * self.att).shape)\n",
    "\n",
    "        alpha = F.leaky_relu(alpha, self.negative_slope)\n",
    "        alpha = softmax(alpha, edge_index_i, ptr, size_i)\n",
    "        self.alpha = alpha\n",
    "\n",
    "        # Sample attention coefficients stochastically.\n",
    "        alpha = F.dropout(alpha, p=self.dropout, training=self.training)\n",
    "        # print(alpha.shape)\n",
    "\n",
    "        return x_j * alpha.view(edge_index_i.shape[0], -1, self.heads, 1)\n",
    "\n",
    "    def update(self, aggr_out):\n",
    "        if self.concat is True:\n",
    "            aggr_out = aggr_out.view(-1, self.heads * self.out_channels)\n",
    "        else:\n",
    "            aggr_out = aggr_out.mean(dim=2)\n",
    "\n",
    "        if self.bias is not None:\n",
    "            aggr_out = aggr_out + self.bias\n",
    "        return aggr_out\n",
    "\n",
    "    def __repr__(self):\n",
    "        return '{}({}, {}, heads={})'.format(self.__class__.__name__,\n",
    "                                             self.in_channels,\n",
    "                                             self.out_channels, self.heads)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2e8c7919",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Net(nn.Module):\n",
    "    def __init__(self, genes, num_peaks, num_tf, cut, mask, num_gos):\n",
    "        super(Net, self).__init__()\n",
    "        self.num_peaks = num_peaks\n",
    "        self.num_tf = num_tf\n",
    "        self.genes = genes\n",
    "        self.cut = cut\n",
    "        self.mask = mask\n",
    "        self.num_gos = num_gos\n",
    "        self.go_dim=6\n",
    "               \n",
    "        self.subnet_modules = nn.ModuleList()\n",
    "        for i in range(len(self.genes)):\n",
    "            gene = self.genes[i]\n",
    "            num_peak = self.num_peaks[i]\n",
    "            self.subnet = subNet(num_peak, self.num_tf)\n",
    "            self.subnet_modules.append(self.subnet)\n",
    "                    \n",
    "        self.cat_activate = nn.LeakyReLU()\n",
    "        self.fc1_weight = nn.Parameter(torch.Tensor(self.num_gos*self.go_dim, len(self.genes)), requires_grad=True)\n",
    "        self.mask_rep = nn.Parameter(torch.repeat_interleave(mask, self.go_dim, dim=0), requires_grad=False)\n",
    "        self.fc1_bias = nn.Parameter(torch.Tensor(self.num_gos*self.go_dim), requires_grad=True)\n",
    "        self.fc1_activate = nn.LeakyReLU()\n",
    "        #self.fc1_drop = nn.Dropout()\n",
    "        self.gat = selfGATConv(self.go_dim, 4, heads=2)\n",
    "        self.gat_activate = nn.LeakyReLU()\n",
    "        self.gat_drop = nn.Dropout()\n",
    "        self.out = nn.Linear(self.num_gos*4, 7)\n",
    "        self.reset_parameters()\n",
    "        \n",
    "    def reset_parameters(self):\n",
    "        nn.init.kaiming_uniform_(self.fc1_weight, a=math.sqrt(5))\n",
    "        fan_in, _ = nn.init._calculate_fan_in_and_fan_out(self.fc1_weight)\n",
    "        bound = 1 / math.sqrt(fan_in)\n",
    "        nn.init.uniform_(self.fc1_bias, -bound, bound)\n",
    "        \n",
    "    def forward(self, x):\n",
    "        x_cat = torch.zeros(x.shape[0], 0).to(device)\n",
    "        for i in range(len(self.subnet_modules)):\n",
    "            x_sub = x[:, :, self.cut[i]:self.cut[i+1]]\n",
    "            x_sub = self.subnet_modules[i](x_sub)\n",
    "            x_cat = torch.cat((x_cat, x_sub), dim=1)\n",
    "\n",
    "        x = self.cat_activate(x_cat)\n",
    "        x = x.matmul((self.fc1_weight*self.mask_rep).t()) + self.fc1_bias\n",
    "        x = self.fc1_activate(x)\n",
    "        #x = self.fc1_drop(x)\n",
    "        x = x.reshape(x.shape[0], self.go_dim, -1)\n",
    "        x = x.permute(2, 0, 1)\n",
    "        x = self.gat(x, edge)\n",
    "        x = x.permute(1, 2, 0)\n",
    "        x = x.reshape(x.shape[0], -1) \n",
    "        x = self.gat_activate(x)\n",
    "        x = self.gat_drop(x)\n",
    "        out = self.out(x)\n",
    "        return x_cat, x, out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "07398bda",
   "metadata": {},
   "outputs": [],
   "source": [
    "def train(model, device, train_loader, optimizer, epoch):\n",
    "    model.train()\n",
    "    #train_loss = 0\n",
    "    for batch_idx, (data, expr, target) in enumerate(train_loader):\n",
    "        data, expr, target = data.to(device), expr.to(device), target.to(device)\n",
    "        optimizer.zero_grad()\n",
    "        expr_hat, cluster_repr, output = model(data)\n",
    "        loss_out = out_criterion(output[:, :6], target[:, :6])\n",
    "        loss_pseudo = expr_criterion(output[:, -1], target[:, -1])\n",
    "        loss_expr = expr_criterion(expr_hat, expr)\n",
    "        loss = 0.05*loss_out + 1*loss_pseudo\n",
    "        #loss = loss_pseudo\n",
    "        loss.backward()\n",
    "        optimizer.step()\n",
    "        #model.fc1.weight.data = model.fc1.weight.mul(torch.repeat_interleave(mask.to(device), 4, dim=0))\n",
    "        if batch_idx % batchsize == 0:\n",
    "            print('\\nTrain Epoch: {} [{}/{} ({:.0f}%)], Total loss: {:.6f}, Expr loss: {:.6f}, Cluster loss: {:.6f}, Pseudo loss: {:.6f}'.\n",
    "                  format(\n",
    "                epoch, batch_idx * len(data), len(train_loader.dataset),\n",
    "                100. * batch_idx / len(train_loader), loss.item(), loss_expr.item(), loss_out.item(), loss_pseudo.item()))\n",
    "        #return(train_loss)\n",
    "\n",
    "                  \n",
    "def test(model, device, test_loader, num_clusters):\n",
    "    model.eval()\n",
    "    with torch.no_grad():\n",
    "        outputs = torch.zeros(0, num_clusters).to(device)\n",
    "        targets = torch.zeros(0).to(device)\n",
    "        outputs_pseudo = torch.zeros(0).to(device)\n",
    "        targets_pseudo = torch.zeros(0).to(device)\n",
    "        test_loss = 0\n",
    "        for data, expr, target in test_loader:\n",
    "            data, expr, target = data.to(device), expr.to(device), target.to(device)\n",
    "            expr_hat, cluster_repr, output = model(data)\n",
    "            \n",
    "            loss_out = out_criterion(output[:, :6], target[:, :6])\n",
    "            loss_pseudo = expr_criterion(output[:, -1], target[:, -1])\n",
    "            loss_expr = expr_criterion(expr_hat, expr)\n",
    "            loss = 0.05*loss_out + 1*loss_pseudo\n",
    "            #loss = loss_pseudo\n",
    "            test_loss = test_loss+loss.item()\n",
    "\n",
    "            target_cluster = target[:, :6].argmax(dim=1)\n",
    "            output_cluster = output[:, :6].softmax(dim=1)\n",
    "            outputs = torch.cat((outputs, output_cluster), dim=0)\n",
    "            targets = torch.cat((targets, target_cluster), dim=0)\n",
    "            target_pseudo = target[:, -1]\n",
    "            output_pseudo = output[:, -1]\n",
    "            outputs_pseudo = torch.cat((outputs_pseudo, output_pseudo), dim=0)\n",
    "            targets_pseudo = torch.cat((targets_pseudo, target_pseudo), dim=0)\n",
    "        f1_score = multiclass_f1_score(outputs, targets, num_classes=num_clusters)\n",
    "        pearsonr, _ = stats.pearsonr(targets_pseudo.detach().cpu().numpy(), outputs_pseudo.detach().cpu().numpy())\n",
    "\n",
    "    return(f1_score, pearsonr, test_loss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c5f58d50",
   "metadata": {},
   "outputs": [],
   "source": [
    "def correlation_score(y_true, y_pred):\n",
    "    #print(np.corrcoef(y_true, y_pred))\n",
    "    return np.corrcoef(y_true, y_pred)[1, 0]\n",
    "\n",
    "def spearman_correlation(y_true, y_pred):\n",
    "    statistic, pvalue = stats.spearmanr(y_true, y_pred)\n",
    "    return abs(statistic)\n",
    "\n",
    "def pearson_correlation(y_true, y_pred):\n",
    "    statistic, pvalue = stats.pearsonr(y_true, y_pred)\n",
    "    return abs(statistic[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d26a9d5",
   "metadata": {},
   "source": [
    "# in-silico perturbation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "4e695a2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "from captum.attr import visualization as viz\n",
    "from captum.attr import Lime, LimeBase, DeepLift, IntegratedGradients, GradientShap, NoiseTunnel, FeatureAblation, KernelShap\n",
    "from captum._utils.models.linear_model import SkLearnLinearRegression, SkLearnLasso"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b17ee15b",
   "metadata": {},
   "outputs": [],
   "source": [
    "use_cuda = True\n",
    "device = torch.device(\"cuda:3\" if use_cuda else \"cpu\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a7d1e25e",
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_x = []\n",
    "edge_y = []\n",
    "for i in range(len(go_filtered)):\n",
    "    edge_x.append(gos.index(go_filtered['id1'].iloc[i]))\n",
    "    edge_y.append(gos.index(go_filtered['id2'].iloc[i]))\n",
    "edge_x = np.array(edge_x)\n",
    "edge_y = np.array(edge_y)\n",
    "edge = np.vstack([edge_x, edge_y])\n",
    "edge = torch.from_numpy(edge).to(device)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "c739b44e",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_tensor = torch.from_numpy(X).float().to(device)#.detach()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e3c6836",
   "metadata": {},
   "source": [
    "## TF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52889adf",
   "metadata": {},
   "source": [
    "### in-silico upregulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "b2881906",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "repeats = 10\n",
    "num_clusters = 6\n",
    "results = np.zeros([repeats, X_tensor.shape[0], X_tensor.shape[1], num_clusters])\n",
    "results_orig = np.zeros([repeats, X_tensor.shape[0], X_tensor.shape[1], num_clusters])\n",
    "results_perturb = np.zeros([repeats, X_tensor.shape[0], X_tensor.shape[1], num_clusters])\n",
    "for i in range(repeats):\n",
    "    model = Net(genes=genes_filtered, num_peaks=num_peaks, num_tf=X.shape[1], cut=cut, mask=mask, num_gos=W_true.shape[0]).to(device)\n",
    "    model.eval()\n",
    "    model.load_state_dict(torch.load('./predict_lineage_pseudotime/model.mask_GAT.leaky.lr001.loss_0.05_1.'+str(i)+'.pt', \n",
    "                                    map_location=device))\n",
    "    np.random.seed(123)  \n",
    "\n",
    "    _, _, orig = model(X_tensor)\n",
    "    orig = F.softmax(orig[:, :6])\n",
    "    for j in range(X_tensor.shape[1]):\n",
    "        X_new = torch.clone(X_tensor).detach()\n",
    "        \n",
    "        tf_mat = cell_by_tf[:, j] + np.quantile(cell_by_tf[:, j], 0.9)\n",
    "        tf_mat = np.repeat(tf_mat[:, np.newaxis], X_tensor.shape[2], axis=1)\n",
    "        peak_mat = cell_by_peak\n",
    "        w = W[j, :]\n",
    "        w = np.repeat(w[np.newaxis, :], X_tensor.shape[0], axis=0)\n",
    "        new = tf_mat*peak_mat*w\n",
    "        \n",
    "        X_new[:, j, :] = torch.from_numpy(new).float().to(device)\n",
    "        _, _, perturb = model(X_new)\n",
    "        perturb = F.softmax(perturb[:, :6])\n",
    "        results[i, :, j, :] = (perturb-orig).detach().cpu().numpy()\n",
    "        results_orig[i, :, j, :] = orig.clone().detach().cpu().numpy()\n",
    "        results_perturb[i, :, j, :] = perturb.detach().cpu().numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "acc9304b",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save('./predict_lineage_pseudotime/interpret_GAT/input_out_TF_0.9quantile.npy', results)\n",
    "np.save('./predict_lineage_pseudotime/interpret_GAT/input_out_TF_0.9quantile_orig.npy', results_orig)\n",
    "np.save('./predict_lineage_pseudotime/interpret_GAT/input_out_TF_0.9quantile_perturb.npy', results_perturb)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "628b07ce",
   "metadata": {},
   "source": [
    "### in-silico knockout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "52684f35",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "repeats = 10\n",
    "num_clusters = 6\n",
    "results = np.zeros([repeats, X_tensor.shape[0], X_tensor.shape[1], num_clusters])\n",
    "results_orig = np.zeros([repeats, X_tensor.shape[0], X_tensor.shape[1], num_clusters])\n",
    "results_perturb = np.zeros([repeats, X_tensor.shape[0], X_tensor.shape[1], num_clusters])\n",
    "for i in range(repeats):\n",
    "    model = Net(genes=genes_filtered, num_peaks=num_peaks, num_tf=X.shape[1], cut=cut, mask=mask, num_gos=W_true.shape[0]).to(device)\n",
    "    model.eval()\n",
    "    model.load_state_dict(torch.load('./predict_lineage_pseudotime/model.mask_GAT.leaky.lr001.loss_0.05_1.'+str(i)+'.pt', \n",
    "                                    map_location=device))\n",
    "    np.random.seed(123)  \n",
    "\n",
    "    _, _, orig = model(X_tensor)\n",
    "    orig = F.softmax(orig[:, :6])\n",
    "    for j in range(X_tensor.shape[1]):\n",
    "        X_new = torch.clone(X_tensor).detach()\n",
    "        \n",
    "        X_new[:, j, :] = 0\n",
    "        _, _, perturb = model(X_new)\n",
    "        perturb = F.softmax(perturb[:, :6])\n",
    "        results[i, :, j, :] = (perturb-orig).detach().cpu().numpy()\n",
    "        results_orig[i, :, j, :] = orig.clone().detach().cpu().numpy()\n",
    "        results_perturb[i, :, j, :] = perturb.detach().cpu().numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "2157dcf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save('./predict_lineage_pseudotime/interpret_GAT/input_out_TF*0.npy', results)\n",
    "np.save('./predict_lineage_pseudotime/interpret_GAT/input_out_TF*0_orig.npy', results_orig)\n",
    "np.save('./predict_lineage_pseudotime/interpret_GAT/input_out_TF*0_perturb.npy', results_perturb)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyenv",
   "language": "python",
   "name": "pyenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
