{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "98db3cb5",
   "metadata": {},
   "source": [
    "This script describes how we filtered candidate pcSNPs from previous GWAS results of T2D. We defined two types of pcSNP:\n",
    "+ pcSNPs that function through regulating the expression of pdTFs;\n",
    "+ pcSNPs that function through regulating the chromatin accessibility of pdCREs.\n",
    "\n",
    "We filtered these two types of pcSNPs and their downstream regulators accordingly."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fa52f48",
   "metadata": {},
   "source": [
    "**Note:**\n",
    "\n",
    "Before running the following scripts, you need to download genomic annotations of the significant GWAS SNPs from the UCSC genome browser or biomart. Here, we provide a processed file \"GWAS_T2D_hg19_UCSC.csv\" in the \"data\" folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bd0d889",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pyfaidx import Fasta\n",
    "from verstack import stratified_continuous_split\n",
    "import matplotlib.pyplot as plt\n",
    "import time\n",
    "import math\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import average_precision_score\n",
    "from sklearn.preprocessing import StandardScaler, MinMaxScaler\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.metrics import mean_squared_error\n",
    "from sklearn.metrics import f1_score\n",
    "from sklearn import preprocessing\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "import scipy\n",
    "from scipy import stats\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "import torch.optim as optim\n",
    "from torch import Tensor\n",
    "from torch.utils.data import TensorDataset, DataLoader\n",
    "from torch.optim.lr_scheduler import StepLR\n",
    "from torch.utils.data import Dataset\n",
    "from torchvision import datasets\n",
    "from torcheval.metrics.functional import multiclass_f1_score\n",
    "from torch_geometric.nn import GCNConv\n",
    "\n",
    "import h5py\n",
    "import seaborn as sns\n",
    "import os\n",
    "os.chdir('/nfs/public/xixi/scRegulate/T2D')\n",
    "\n",
    "from typing import Tuple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "950b3dc7",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "df_x = pd.read_csv('./data/beta.atac.aggregate_30cells.csv', index_col=0).transpose()\n",
    "df_y = pd.read_csv('./data/beta.rna.aggregate_30cells.csv', index_col=0).transpose()\n",
    "df_peaks = pd.DataFrame(df_x.columns)[0].str.split('-',expand=True) \n",
    "df_peaks = df_peaks.rename(columns={0: \"chrom\", 1: \"start\", 2: \"end\"})\n",
    "df_peaks[\"start\"] = pd.to_numeric(df_peaks[\"start\"])\n",
    "df_peaks[\"end\"] = pd.to_numeric(df_peaks[\"end\"])\n",
    "geneanno = pd.read_csv('../../ref_genome/hg19_geneanno.txt', sep='\\t')\n",
    "geneanno = geneanno.drop_duplicates(subset=['Gene name'])\n",
    "motif_files = os.listdir('../../ref_genome/JASPAR_motifs_pfm_homosapiens/pfm.np')\n",
    "tfs_kept = []\n",
    "tf_by_region_mat = []\n",
    "for i in list(motif_files):\n",
    "    tf = i.split('.')[-2]#.capitalize()\n",
    "    if tf in df_y.columns:\n",
    "        if tf in tfs_kept:\n",
    "            continue\n",
    "            \n",
    "        tfs_kept.append(tf)\n",
    "tfs_kept = sorted(tfs_kept)\n",
    "h5f = h5py.File('./predict_status/data_T2D_float16.h5', 'r')\n",
    "#X = h5f['X'][:]\n",
    "expr = h5f['expr'][:]\n",
    "num_peaks = h5f['num_peaks'][:]\n",
    "peaks = pd.DataFrame(np.array([item.decode('utf-8') for j in h5f['peaks'][:] for item in j]).reshape(-1, 3))\n",
    "h5f.close()\n",
    "cut = [0]\n",
    "s = 0\n",
    "for i in num_peaks:\n",
    "    s = s+i\n",
    "    cut.append(s)\n",
    "y = pd.read_csv('./data/beta.label.aggregate_30cells.csv', index_col=0)\n",
    "\n",
    "y = y['status'].to_list()\n",
    "y_new = y\n",
    "values_to_replace = ['Non-diabetic', 'Pre-T2D', 'T2D']\n",
    "for value in values_to_replace:\n",
    "    y_new = np.where(y == value, 0, y_new)\n",
    "y_new = y_new.reshape((len(y), 1))\n",
    "enc = OneHotEncoder(handle_unknown='ignore')\n",
    "y_oht = enc.fit_transform(y_new).toarray()\n",
    "expr_train, expr_test, y_oht_train, y_oht_test, y_label_train, y_label_test = train_test_split(expr, y_oht, y, test_size=0.3,\n",
    "                                                                                                    random_state=2024, stratify=y)\n",
    "expr_test, expr_val, y_oht_test, y_oht_val, y_label_test, y_label_val = train_test_split(expr_test, y_oht_test, y_label_test, \n",
    "                                                                                                test_size=0.5, random_state=2024, \n",
    "                                                                                            stratify=y_label_test)\n",
    "\n",
    "class TensorDataset(Dataset[Tuple[Tensor, ...]]):\n",
    "    r\"\"\"Dataset wrapping tensors.\n",
    "\n",
    "    Each sample will be retrieved by indexing tensors along the first dimension.\n",
    "\n",
    "    Args:\n",
    "        *tensors (Tensor): tensors that have the same size of the first dimension.\n",
    "    \"\"\"\n",
    "    tensors: Tuple[Tensor, ...]\n",
    "\n",
    "    def __init__(self, *tensors: Tensor) -> None:\n",
    "        assert all(tensors[0].size(0) == tensor.size(0) for tensor in tensors), \"Size mismatch between tensors\"\n",
    "        self.tensors = tensors\n",
    "\n",
    "    def __getitem__(self, index):\n",
    "        return tuple(tensor[index] for tensor in self.tensors)\n",
    "\n",
    "    def __len__(self):\n",
    "        return self.tensors[0].size(0)\n",
    "\n",
    "class subNet(nn.Module):\n",
    "    def __init__(self, num_peak, num_tf):\n",
    "        super(subNet, self).__init__()\n",
    "        self.num_peak = num_peak\n",
    "        self.num_tf = num_tf\n",
    "\n",
    "        self.fc1 = nn.Linear(self.num_peak * self.num_tf, 1)\n",
    "        self.fc1_activate = nn.ReLU()\n",
    "        self.abs = nn.ReLU()\n",
    "        \n",
    "    def forward(self, x):\n",
    "        x = x.reshape(x.shape[0], -1)\n",
    "        x = self.fc1(x)\n",
    "        return x\n",
    "\n",
    "class Net(nn.Module):\n",
    "    def __init__(self, num_genes, num_peaks, num_tf, cut):\n",
    "        super(Net, self).__init__()\n",
    "        self.num_peaks = num_peaks\n",
    "        self.num_tf = num_tf\n",
    "        self.num_genes = num_genes\n",
    "        self.cut = cut\n",
    "        self.gene_dim = 2\n",
    "               \n",
    "        self.subnet_modules = nn.ModuleList()\n",
    "        for i in range(num_genes):\n",
    "            num_peak = self.num_peaks[i]\n",
    "            self.subnet = subNet(num_peak, self.num_tf)\n",
    "            self.subnet_modules.append(self.subnet)\n",
    "                    \n",
    "        self.cat_activate = nn.ReLU()\n",
    "        self.conv = GCNConv(1, self.gene_dim, add_self_loops=False)\n",
    "        #self.fc1 = nn.Linear(self.num_genes, 100)\n",
    "        self.conv_activate = nn.ReLU()\n",
    "        self.out = nn.Linear(self.num_genes*self.gene_dim, 3)\n",
    "        \n",
    "        #self.initialize_parameters()\n",
    "        \n",
    "    def initialize_parameters(self):\n",
    "        weight = self.conv.lin.weight\n",
    "        bias = self.conv.bias\n",
    "        nn.init.kaiming_uniform_(weight, a=math.sqrt(5))\n",
    "        fan_in, _ = nn.init._calculate_fan_in_and_fan_out(weight)\n",
    "        #torch.nn.init.xavier_uniform_(weight)\n",
    "        if bias is not None:\n",
    "            bound = 1 / math.sqrt(fan_in)\n",
    "            nn.init.uniform_(bias, -bound, bound)\n",
    "\n",
    "    def forward(self, x):\n",
    "        x_cat = torch.zeros(x.shape[0], 0).to(device)\n",
    "        for i in range(len(self.subnet_modules)):\n",
    "            x_sub = x[:, :, self.cut[i]:self.cut[i+1]]\n",
    "            x_sub = self.subnet_modules[i](x_sub)\n",
    "            x_cat = torch.cat((x_cat, x_sub), dim=1)\n",
    "\n",
    "        x_cat = self.cat_activate(x_cat)\n",
    "        x = torch.unsqueeze(x_cat, 2)\n",
    "        #x = F.dropout(x, p=0.3)\n",
    "        x = self.conv_activate(self.conv(x, edge))\n",
    "        #x = F.dropout(x, p=0.3)\n",
    "        x = x.reshape(x.shape[0], -1)\n",
    "        out = self.out(x)\n",
    "        return x_cat, x, out\n",
    "\n",
    "def train(model, device, train_loader, optimizer, epoch):\n",
    "    model.train()\n",
    "    #train_loss = 0\n",
    "    for batch_idx, (data, expr, target) in enumerate(train_loader):\n",
    "        data, expr, target = data.float(), expr.float(), target.float()\n",
    "        optimizer.zero_grad()\n",
    "        expr_hat, cluster_repr, output = model(data)\n",
    "        loss_out = out_criterion(output, target)\n",
    "        #loss_pseudo = expr_criterion(output[:, -1], target[:, -1])\n",
    "        loss_expr = expr_criterion(expr_hat, expr)\n",
    "        loss = loss_out #+ 0.7*loss_pseudo\n",
    "        #loss = loss_pseudo\n",
    "        loss.backward()\n",
    "        optimizer.step()\n",
    "        #model.fc1.weight.data = model.fc1.weight.mul(torch.repeat_interleave(mask.to(device), 4, dim=0))\n",
    "#         if batch_idx % batchsize == 0:\n",
    "#             print('\\nTrain Epoch: {} [{}/{} ({:.0f}%)], Expr loss: {:.6f}, Cluster loss: {:.6f}'.\n",
    "#                   format(\n",
    "#                 epoch, batch_idx * len(data), len(train_loader.dataset),\n",
    "#                 100. * batch_idx / len(train_loader), loss_expr.item(), loss_out.item()))\n",
    "        #return(train_loss)\n",
    "\n",
    "                  \n",
    "def test(model, device, test_loader, num_clusters):\n",
    "    model.eval()\n",
    "    with torch.no_grad():\n",
    "        outputs = torch.zeros(0, num_clusters).to(device)\n",
    "        targets = torch.zeros(0).to(device)\n",
    "        outputs_pseudo = torch.zeros(0).to(device)\n",
    "        targets_pseudo = torch.zeros(0).to(device)\n",
    "        test_loss = 0\n",
    "        for data, expr, target in test_loader:\n",
    "            data, expr, target = data.float(), expr.float(), target.float()\n",
    "            expr_hat, cluster_repr, output = model(data)\n",
    "            \n",
    "            loss_out = out_criterion(output, target)\n",
    "            #loss_pseudo = expr_criterion(output[:, -1], target[:, -1])\n",
    "            loss_expr = expr_criterion(expr_hat, expr)\n",
    "            loss = loss_out\n",
    "            #loss = loss_pseudo\n",
    "            test_loss = test_loss+loss.item()\n",
    "\n",
    "            target_cluster = target.argmax(dim=1)\n",
    "            output_cluster = output.softmax(dim=1)\n",
    "            outputs = torch.cat((outputs, output_cluster), dim=0)\n",
    "            targets = torch.cat((targets, target_cluster), dim=0)\n",
    "            #target_pseudo = target[:, -1]\n",
    "            #output_pseudo = output[:, -1]\n",
    "            #outputs_pseudo = torch.cat((outputs_pseudo, output_pseudo), dim=0)\n",
    "            #targets_pseudo = torch.cat((targets_pseudo, target_pseudo), dim=0)\n",
    "        f1_score = multiclass_f1_score(outputs, targets, num_classes=num_clusters)\n",
    "        #pearsonr, _ = stats.pearsonr(targets_pseudo.detach().cpu().numpy(), outputs_pseudo.detach().cpu().numpy())\n",
    "\n",
    "    return(f1_score, test_loss)\n",
    "\n",
    "def correlation_score(y_true, y_pred):\n",
    "    #print(np.corrcoef(y_true, y_pred))\n",
    "    return np.corrcoef(y_true, y_pred)[1, 0]\n",
    "\n",
    "def spearman_correlation(y_true, y_pred):\n",
    "    statistic, pvalue = stats.spearmanr(y_true, y_pred)\n",
    "    return abs(statistic)\n",
    "\n",
    "def pearson_correlation(y_true, y_pred):\n",
    "    statistic, pvalue = stats.pearsonr(y_true, y_pred)\n",
    "    return abs(statistic[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ad256b9",
   "metadata": {},
   "source": [
    "# pcSNP -> pdTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "02b12886",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GLIS3', 'RORA', 'RFX3', 'MEIS2', 'FOS', 'HES1', 'FOXP2', 'MLXIPL', 'ETV1', 'NR1D2', 'TRPS1', 'HNF4A', 'PBX3', 'BACH2', 'BACH1', 'TCF4', 'JUND', 'TEAD1', 'KLF6', 'ZKSCAN1']\n",
      "['GLIS3', 'RFX3', 'RORA', 'MEIS2', 'FOXP2', 'ETV1', 'NR1D2', 'HES1', 'TRPS1', 'BACH1', 'FOS', 'MLXIPL', 'BACH2', 'TCF4', 'PBX3', 'TEAD1', 'ETV5', 'CUX1', 'KLF6', 'FOXO3']\n"
     ]
    }
   ],
   "source": [
    "results = np.load('./predict_status/interpret/input_output_TF_0.9quantile.GCN.string.npy')\n",
    "import scipy.stats as ss\n",
    "rank_all = []\n",
    "for rep in range(10):\n",
    "    #top = int(0.1*len(tfs_kept))\n",
    "    cells = results[rep, :, :, :]\n",
    "    rank = ss.rankdata(np.array([np.sum(abs(cells[:, i, :])) for i in range(len(tfs_kept))])) #rank from the smallest to the biggest\n",
    "    rank_all.append(rank)\n",
    "\n",
    "rank_all = np.array(rank_all)\n",
    "top = 20\n",
    "top_tfs_ind = np.argsort(np.mean(rank_all, axis=0))[::-1][:top]  # rank from the biggest to the smallest\n",
    "print([tfs_kept[i] for i in top_tfs_ind])\n",
    "top_over = [tfs_kept[i] for i in top_tfs_ind]\n",
    "\n",
    "results = np.load('./predict_status/interpret/input_output_TF*0.GCN.string.npy')\n",
    "import scipy.stats as ss\n",
    "rank_all = []\n",
    "for rep in range(10):\n",
    "    #top = int(0.1*len(tfs_kept))\n",
    "    cells = results[rep, :, :, :]\n",
    "    rank = ss.rankdata(np.array([np.sum(abs(cells[:, i, :])) for i in range(len(tfs_kept))])) #rank from the smallest to the biggest\n",
    "    rank_all.append(rank)\n",
    "\n",
    "rank_all = np.array(rank_all)\n",
    "top = 20\n",
    "top_tfs_ind0 = np.argsort(np.mean(rank_all, axis=0))[::-1][:top]  # rank from the biggest to the smallest\n",
    "print([tfs_kept[i] for i in top_tfs_ind0])\n",
    "top_knockout = [tfs_kept[i] for i in top_tfs_ind0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "750f6226",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['FOXO3', 'RFX3', 'HNF4A', 'FOXP2', 'PBX3', 'BACH2', 'JUND', 'FOS', 'ZKSCAN1', 'MEIS2', 'ETV1', 'TCF4', 'TEAD1', 'CUX1', 'RORA', 'ETV5', 'TRPS1', 'HES1', 'KLF6', 'MLXIPL', 'NR1D2', 'BACH1', 'GLIS3']\n"
     ]
    }
   ],
   "source": [
    "tfs_selected = list(set(top_over + top_knockout))\n",
    "print(tfs_selected)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "028cfbf4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "23"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(tfs_selected)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "adfc604d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Transcription start site (TSS)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13069</th>\n",
       "      <td>ENSG00000067082</td>\n",
       "      <td>ENST00000461124</td>\n",
       "      <td>10</td>\n",
       "      <td>3818188</td>\n",
       "      <td>3827473</td>\n",
       "      <td>-1</td>\n",
       "      <td>KLF6</td>\n",
       "      <td>3822409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24038</th>\n",
       "      <td>ENSG00000187079</td>\n",
       "      <td>ENST00000527636</td>\n",
       "      <td>11</td>\n",
       "      <td>12695969</td>\n",
       "      <td>12966298</td>\n",
       "      <td>1</td>\n",
       "      <td>TEAD1</td>\n",
       "      <td>12696138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49405</th>\n",
       "      <td>ENSG00000170345</td>\n",
       "      <td>ENST00000303562</td>\n",
       "      <td>14</td>\n",
       "      <td>75745477</td>\n",
       "      <td>75748933</td>\n",
       "      <td>1</td>\n",
       "      <td>FOS</td>\n",
       "      <td>75745477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52928</th>\n",
       "      <td>ENSG00000134138</td>\n",
       "      <td>ENST00000338564</td>\n",
       "      <td>15</td>\n",
       "      <td>37181406</td>\n",
       "      <td>37393504</td>\n",
       "      <td>-1</td>\n",
       "      <td>MEIS2</td>\n",
       "      <td>37393406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55241</th>\n",
       "      <td>ENSG00000069667</td>\n",
       "      <td>ENST00000335670</td>\n",
       "      <td>15</td>\n",
       "      <td>60780483</td>\n",
       "      <td>61521518</td>\n",
       "      <td>-1</td>\n",
       "      <td>RORA</td>\n",
       "      <td>61521518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83979</th>\n",
       "      <td>ENSG00000196628</td>\n",
       "      <td>ENST00000356073</td>\n",
       "      <td>18</td>\n",
       "      <td>52889562</td>\n",
       "      <td>53332018</td>\n",
       "      <td>-1</td>\n",
       "      <td>TCF4</td>\n",
       "      <td>53255860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89675</th>\n",
       "      <td>ENSG00000130522</td>\n",
       "      <td>ENST00000600972</td>\n",
       "      <td>19</td>\n",
       "      <td>18390563</td>\n",
       "      <td>18392432</td>\n",
       "      <td>-1</td>\n",
       "      <td>JUND</td>\n",
       "      <td>18391739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113343</th>\n",
       "      <td>ENSG00000101076</td>\n",
       "      <td>ENST00000316673</td>\n",
       "      <td>20</td>\n",
       "      <td>42984340</td>\n",
       "      <td>43061485</td>\n",
       "      <td>1</td>\n",
       "      <td>HNF4A</td>\n",
       "      <td>42984340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114053</th>\n",
       "      <td>ENSG00000156273</td>\n",
       "      <td>ENST00000548219</td>\n",
       "      <td>21</td>\n",
       "      <td>30566392</td>\n",
       "      <td>31003071</td>\n",
       "      <td>1</td>\n",
       "      <td>BACH1</td>\n",
       "      <td>30566392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123007</th>\n",
       "      <td>ENSG00000244405</td>\n",
       "      <td>ENST00000306376</td>\n",
       "      <td>3</td>\n",
       "      <td>185764097</td>\n",
       "      <td>185828107</td>\n",
       "      <td>-1</td>\n",
       "      <td>ETV5</td>\n",
       "      <td>185826901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127607</th>\n",
       "      <td>ENSG00000174738</td>\n",
       "      <td>ENST00000312521</td>\n",
       "      <td>3</td>\n",
       "      <td>23986751</td>\n",
       "      <td>24022109</td>\n",
       "      <td>1</td>\n",
       "      <td>NR1D2</td>\n",
       "      <td>23986751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130449</th>\n",
       "      <td>ENSG00000114315</td>\n",
       "      <td>ENST00000232424</td>\n",
       "      <td>3</td>\n",
       "      <td>193853934</td>\n",
       "      <td>193856521</td>\n",
       "      <td>1</td>\n",
       "      <td>HES1</td>\n",
       "      <td>193853934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152238</th>\n",
       "      <td>ENSG00000112182</td>\n",
       "      <td>ENST00000257749</td>\n",
       "      <td>6</td>\n",
       "      <td>90636248</td>\n",
       "      <td>91006627</td>\n",
       "      <td>-1</td>\n",
       "      <td>BACH2</td>\n",
       "      <td>91006461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152633</th>\n",
       "      <td>ENSG00000118689</td>\n",
       "      <td>ENST00000343882</td>\n",
       "      <td>6</td>\n",
       "      <td>108881038</td>\n",
       "      <td>109005977</td>\n",
       "      <td>1</td>\n",
       "      <td>FOXO3</td>\n",
       "      <td>108881038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156240</th>\n",
       "      <td>ENSG00000006468</td>\n",
       "      <td>ENST00000430479</td>\n",
       "      <td>7</td>\n",
       "      <td>13930853</td>\n",
       "      <td>14031050</td>\n",
       "      <td>-1</td>\n",
       "      <td>ETV1</td>\n",
       "      <td>14029571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159091</th>\n",
       "      <td>ENSG00000009950</td>\n",
       "      <td>ENST00000414749</td>\n",
       "      <td>7</td>\n",
       "      <td>73007524</td>\n",
       "      <td>73038873</td>\n",
       "      <td>-1</td>\n",
       "      <td>MLXIPL</td>\n",
       "      <td>73038822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160831</th>\n",
       "      <td>ENSG00000106261</td>\n",
       "      <td>ENST00000324306</td>\n",
       "      <td>7</td>\n",
       "      <td>99613204</td>\n",
       "      <td>99639312</td>\n",
       "      <td>1</td>\n",
       "      <td>ZKSCAN1</td>\n",
       "      <td>99613204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161354</th>\n",
       "      <td>ENSG00000257923</td>\n",
       "      <td>ENST00000437600</td>\n",
       "      <td>7</td>\n",
       "      <td>101458959</td>\n",
       "      <td>101927249</td>\n",
       "      <td>1</td>\n",
       "      <td>CUX1</td>\n",
       "      <td>101458959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162082</th>\n",
       "      <td>ENSG00000128573</td>\n",
       "      <td>ENST00000412402</td>\n",
       "      <td>7</td>\n",
       "      <td>113726382</td>\n",
       "      <td>114333827</td>\n",
       "      <td>1</td>\n",
       "      <td>FOXP2</td>\n",
       "      <td>113726382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170664</th>\n",
       "      <td>ENSG00000104447</td>\n",
       "      <td>ENST00000395715</td>\n",
       "      <td>8</td>\n",
       "      <td>116420724</td>\n",
       "      <td>116821899</td>\n",
       "      <td>-1</td>\n",
       "      <td>TRPS1</td>\n",
       "      <td>116681228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172694</th>\n",
       "      <td>ENSG00000080298</td>\n",
       "      <td>ENST00000382004</td>\n",
       "      <td>9</td>\n",
       "      <td>3218297</td>\n",
       "      <td>3526004</td>\n",
       "      <td>-1</td>\n",
       "      <td>RFX3</td>\n",
       "      <td>3525983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172712</th>\n",
       "      <td>ENSG00000107249</td>\n",
       "      <td>ENST00000324333</td>\n",
       "      <td>9</td>\n",
       "      <td>3824127</td>\n",
       "      <td>4348392</td>\n",
       "      <td>-1</td>\n",
       "      <td>GLIS3</td>\n",
       "      <td>4152183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177284</th>\n",
       "      <td>ENSG00000167081</td>\n",
       "      <td>ENST00000342287</td>\n",
       "      <td>9</td>\n",
       "      <td>128509624</td>\n",
       "      <td>128729656</td>\n",
       "      <td>1</td>\n",
       "      <td>PBX3</td>\n",
       "      <td>128509624</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene stable ID Transcript stable ID  Chromosome/scaffold name  \\\n",
       "13069   ENSG00000067082      ENST00000461124                        10   \n",
       "24038   ENSG00000187079      ENST00000527636                        11   \n",
       "49405   ENSG00000170345      ENST00000303562                        14   \n",
       "52928   ENSG00000134138      ENST00000338564                        15   \n",
       "55241   ENSG00000069667      ENST00000335670                        15   \n",
       "83979   ENSG00000196628      ENST00000356073                        18   \n",
       "89675   ENSG00000130522      ENST00000600972                        19   \n",
       "113343  ENSG00000101076      ENST00000316673                        20   \n",
       "114053  ENSG00000156273      ENST00000548219                        21   \n",
       "123007  ENSG00000244405      ENST00000306376                         3   \n",
       "127607  ENSG00000174738      ENST00000312521                         3   \n",
       "130449  ENSG00000114315      ENST00000232424                         3   \n",
       "152238  ENSG00000112182      ENST00000257749                         6   \n",
       "152633  ENSG00000118689      ENST00000343882                         6   \n",
       "156240  ENSG00000006468      ENST00000430479                         7   \n",
       "159091  ENSG00000009950      ENST00000414749                         7   \n",
       "160831  ENSG00000106261      ENST00000324306                         7   \n",
       "161354  ENSG00000257923      ENST00000437600                         7   \n",
       "162082  ENSG00000128573      ENST00000412402                         7   \n",
       "170664  ENSG00000104447      ENST00000395715                         8   \n",
       "172694  ENSG00000080298      ENST00000382004                         9   \n",
       "172712  ENSG00000107249      ENST00000324333                         9   \n",
       "177284  ENSG00000167081      ENST00000342287                         9   \n",
       "\n",
       "        Gene start (bp)  Gene end (bp)  Strand Gene name  \\\n",
       "13069           3818188        3827473      -1      KLF6   \n",
       "24038          12695969       12966298       1     TEAD1   \n",
       "49405          75745477       75748933       1       FOS   \n",
       "52928          37181406       37393504      -1     MEIS2   \n",
       "55241          60780483       61521518      -1      RORA   \n",
       "83979          52889562       53332018      -1      TCF4   \n",
       "89675          18390563       18392432      -1      JUND   \n",
       "113343         42984340       43061485       1     HNF4A   \n",
       "114053         30566392       31003071       1     BACH1   \n",
       "123007        185764097      185828107      -1      ETV5   \n",
       "127607         23986751       24022109       1     NR1D2   \n",
       "130449        193853934      193856521       1      HES1   \n",
       "152238         90636248       91006627      -1     BACH2   \n",
       "152633        108881038      109005977       1     FOXO3   \n",
       "156240         13930853       14031050      -1      ETV1   \n",
       "159091         73007524       73038873      -1    MLXIPL   \n",
       "160831         99613204       99639312       1   ZKSCAN1   \n",
       "161354        101458959      101927249       1      CUX1   \n",
       "162082        113726382      114333827       1     FOXP2   \n",
       "170664        116420724      116821899      -1     TRPS1   \n",
       "172694          3218297        3526004      -1      RFX3   \n",
       "172712          3824127        4348392      -1     GLIS3   \n",
       "177284        128509624      128729656       1      PBX3   \n",
       "\n",
       "        Transcription start site (TSS)  \n",
       "13069                          3822409  \n",
       "24038                         12696138  \n",
       "49405                         75745477  \n",
       "52928                         37393406  \n",
       "55241                         61521518  \n",
       "83979                         53255860  \n",
       "89675                         18391739  \n",
       "113343                        42984340  \n",
       "114053                        30566392  \n",
       "123007                       185826901  \n",
       "127607                        23986751  \n",
       "130449                       193853934  \n",
       "152238                        91006461  \n",
       "152633                       108881038  \n",
       "156240                        14029571  \n",
       "159091                        73038822  \n",
       "160831                        99613204  \n",
       "161354                       101458959  \n",
       "162082                       113726382  \n",
       "170664                       116681228  \n",
       "172694                         3525983  \n",
       "172712                         4152183  \n",
       "177284                       128509624  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geneanno = pd.read_csv('../../ref_genome/hg19_geneanno.txt', sep='\\t')\n",
    "geneanno = geneanno.drop_duplicates(subset=['Gene name'])\n",
    "geneanno = geneanno[geneanno['Gene name'].isin(tfs_selected)]\n",
    "geneanno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c499d96b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>TSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KLF6</td>\n",
       "      <td>chr10</td>\n",
       "      <td>3572409</td>\n",
       "      <td>4072409</td>\n",
       "      <td>-1</td>\n",
       "      <td>3822409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TEAD1</td>\n",
       "      <td>chr11</td>\n",
       "      <td>12446138</td>\n",
       "      <td>12946138</td>\n",
       "      <td>1</td>\n",
       "      <td>12696138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FOS</td>\n",
       "      <td>chr14</td>\n",
       "      <td>75495477</td>\n",
       "      <td>75995477</td>\n",
       "      <td>1</td>\n",
       "      <td>75745477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MEIS2</td>\n",
       "      <td>chr15</td>\n",
       "      <td>37143406</td>\n",
       "      <td>37643406</td>\n",
       "      <td>-1</td>\n",
       "      <td>37393406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RORA</td>\n",
       "      <td>chr15</td>\n",
       "      <td>61271518</td>\n",
       "      <td>61771518</td>\n",
       "      <td>-1</td>\n",
       "      <td>61521518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TCF4</td>\n",
       "      <td>chr18</td>\n",
       "      <td>53005860</td>\n",
       "      <td>53505860</td>\n",
       "      <td>-1</td>\n",
       "      <td>53255860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>JUND</td>\n",
       "      <td>chr19</td>\n",
       "      <td>18141739</td>\n",
       "      <td>18641739</td>\n",
       "      <td>-1</td>\n",
       "      <td>18391739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HNF4A</td>\n",
       "      <td>chr20</td>\n",
       "      <td>42734340</td>\n",
       "      <td>43234340</td>\n",
       "      <td>1</td>\n",
       "      <td>42984340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BACH1</td>\n",
       "      <td>chr21</td>\n",
       "      <td>30316392</td>\n",
       "      <td>30816392</td>\n",
       "      <td>1</td>\n",
       "      <td>30566392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ETV5</td>\n",
       "      <td>chr3</td>\n",
       "      <td>185576901</td>\n",
       "      <td>186076901</td>\n",
       "      <td>-1</td>\n",
       "      <td>185826901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NR1D2</td>\n",
       "      <td>chr3</td>\n",
       "      <td>23736751</td>\n",
       "      <td>24236751</td>\n",
       "      <td>1</td>\n",
       "      <td>23986751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HES1</td>\n",
       "      <td>chr3</td>\n",
       "      <td>193603934</td>\n",
       "      <td>194103934</td>\n",
       "      <td>1</td>\n",
       "      <td>193853934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>BACH2</td>\n",
       "      <td>chr6</td>\n",
       "      <td>90756461</td>\n",
       "      <td>91256461</td>\n",
       "      <td>-1</td>\n",
       "      <td>91006461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>FOXO3</td>\n",
       "      <td>chr6</td>\n",
       "      <td>108631038</td>\n",
       "      <td>109131038</td>\n",
       "      <td>1</td>\n",
       "      <td>108881038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ETV1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>13779571</td>\n",
       "      <td>14279571</td>\n",
       "      <td>-1</td>\n",
       "      <td>14029571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>MLXIPL</td>\n",
       "      <td>chr7</td>\n",
       "      <td>72788822</td>\n",
       "      <td>73288822</td>\n",
       "      <td>-1</td>\n",
       "      <td>73038822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ZKSCAN1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>99363204</td>\n",
       "      <td>99863204</td>\n",
       "      <td>1</td>\n",
       "      <td>99613204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CUX1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>101208959</td>\n",
       "      <td>101708959</td>\n",
       "      <td>1</td>\n",
       "      <td>101458959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>113476382</td>\n",
       "      <td>113976382</td>\n",
       "      <td>1</td>\n",
       "      <td>113726382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TRPS1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>116431228</td>\n",
       "      <td>116931228</td>\n",
       "      <td>-1</td>\n",
       "      <td>116681228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>RFX3</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3275983</td>\n",
       "      <td>3775983</td>\n",
       "      <td>-1</td>\n",
       "      <td>3525983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>GLIS3</td>\n",
       "      <td>chr9</td>\n",
       "      <td>3902183</td>\n",
       "      <td>4402183</td>\n",
       "      <td>-1</td>\n",
       "      <td>4152183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>PBX3</td>\n",
       "      <td>chr9</td>\n",
       "      <td>128259624</td>\n",
       "      <td>128759624</td>\n",
       "      <td>1</td>\n",
       "      <td>128509624</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene  chrom      start        end  strand        TSS\n",
       "0      KLF6  chr10    3572409    4072409      -1    3822409\n",
       "1     TEAD1  chr11   12446138   12946138       1   12696138\n",
       "2       FOS  chr14   75495477   75995477       1   75745477\n",
       "3     MEIS2  chr15   37143406   37643406      -1   37393406\n",
       "4      RORA  chr15   61271518   61771518      -1   61521518\n",
       "5      TCF4  chr18   53005860   53505860      -1   53255860\n",
       "6      JUND  chr19   18141739   18641739      -1   18391739\n",
       "7     HNF4A  chr20   42734340   43234340       1   42984340\n",
       "8     BACH1  chr21   30316392   30816392       1   30566392\n",
       "9      ETV5   chr3  185576901  186076901      -1  185826901\n",
       "10    NR1D2   chr3   23736751   24236751       1   23986751\n",
       "11     HES1   chr3  193603934  194103934       1  193853934\n",
       "12    BACH2   chr6   90756461   91256461      -1   91006461\n",
       "13    FOXO3   chr6  108631038  109131038       1  108881038\n",
       "14     ETV1   chr7   13779571   14279571      -1   14029571\n",
       "15   MLXIPL   chr7   72788822   73288822      -1   73038822\n",
       "16  ZKSCAN1   chr7   99363204   99863204       1   99613204\n",
       "17     CUX1   chr7  101208959  101708959       1  101458959\n",
       "18    FOXP2   chr7  113476382  113976382       1  113726382\n",
       "19    TRPS1   chr8  116431228  116931228      -1  116681228\n",
       "20     RFX3   chr9    3275983    3775983      -1    3525983\n",
       "21    GLIS3   chr9    3902183    4402183      -1    4152183\n",
       "22     PBX3   chr9  128259624  128759624       1  128509624"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chroms = []\n",
    "starts = []\n",
    "ends = []\n",
    "for i in range(len(geneanno)):\n",
    "    strand = geneanno['Strand'].iloc[i]\n",
    "    chroms.append('chr'+geneanno['Chromosome/scaffold name'].astype(str).iloc[i])\n",
    "    starts.append(geneanno['Transcription start site (TSS)'].iloc[i]-250000)\n",
    "    ends.append(geneanno['Transcription start site (TSS)'].iloc[i]+250000)\n",
    "df_generegions = pd.DataFrame({'gene': geneanno['Gene name'],\n",
    "                             'chrom': chroms,\n",
    "                             'start': starts,\n",
    "                             'end': ends,\n",
    "                              'strand': geneanno['Strand'],\n",
    "                             'TSS': geneanno['Transcription start site (TSS)']})\n",
    "df_generegions = df_generegions.reset_index(drop=True)\n",
    "df_generegions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0a28c110",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>snp_name</th>\n",
       "      <th>ref</th>\n",
       "      <th>altCount</th>\n",
       "      <th>alts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183004333</td>\n",
       "      <td>183004334</td>\n",
       "      <td>rs4129858</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>G,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17</td>\n",
       "      <td>44083947</td>\n",
       "      <td>44083948</td>\n",
       "      <td>rs8067056</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "      <td>C,G,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>51438364</td>\n",
       "      <td>51438365</td>\n",
       "      <td>rs3176466</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>51457199</td>\n",
       "      <td>51457200</td>\n",
       "      <td>rs72906810</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>C,G,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>51506885</td>\n",
       "      <td>51506886</td>\n",
       "      <td>rs12088739</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>G,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2436</th>\n",
       "      <td>chr4</td>\n",
       "      <td>941517</td>\n",
       "      <td>941518</td>\n",
       "      <td>rs2290402</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2437</th>\n",
       "      <td>chr5</td>\n",
       "      <td>36257017</td>\n",
       "      <td>36257018</td>\n",
       "      <td>rs16902871</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>G,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2438</th>\n",
       "      <td>chr11</td>\n",
       "      <td>64100775</td>\n",
       "      <td>64100776</td>\n",
       "      <td>rs1662185</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>G,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2439</th>\n",
       "      <td>chr3</td>\n",
       "      <td>72803589</td>\n",
       "      <td>72803590</td>\n",
       "      <td>rs9814945</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>A,T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2440</th>\n",
       "      <td>chr3</td>\n",
       "      <td>72813581</td>\n",
       "      <td>72813582</td>\n",
       "      <td>rs11922794</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>A,C,</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2441 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  chromStart   chromEnd    snp_name ref  altCount  alts\n",
       "0      chr1   183004333  183004334   rs4129858   A         1    G,\n",
       "1     chr17    44083947   44083948   rs8067056   T         2  C,G,\n",
       "2      chr1    51438364   51438365   rs3176466   C         1    T,\n",
       "3      chr1    51457199   51457200  rs72906810   A         2  C,G,\n",
       "4      chr1    51506885   51506886  rs12088739   A         1    G,\n",
       "...     ...         ...        ...         ...  ..       ...   ...\n",
       "2436   chr4      941517     941518   rs2290402   C         1    T,\n",
       "2437   chr5    36257017   36257018  rs16902871   A         1    G,\n",
       "2438  chr11    64100775   64100776   rs1662185   A         1    G,\n",
       "2439   chr3    72803589   72803590   rs9814945   C         2  A,T,\n",
       "2440   chr3    72813581   72813582  rs11922794   G         2  A,C,\n",
       "\n",
       "[2441 rows x 7 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gwas_hg19 = pd.read_csv('/nfs/public/xixi/scRegulate/GWAS/GWAS_T2D_hg19_UCSC.csv')\n",
    "df_gwas_hg19 = df_gwas_hg19.iloc[:, :7]\n",
    "df_gwas_hg19 = df_gwas_hg19.rename(columns={df_gwas_hg19.columns[3]: 'snp_name'})\n",
    "df_gwas_hg19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "587b62d8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_name</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs72926932</td>\n",
       "      <td>TCF4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs28719468</td>\n",
       "      <td>TCF4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs76811102</td>\n",
       "      <td>HNF4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs6017317</td>\n",
       "      <td>HNF4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs4812829</td>\n",
       "      <td>HNF4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs16988991</td>\n",
       "      <td>HNF4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs113810779</td>\n",
       "      <td>HNF4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rs12625671</td>\n",
       "      <td>HNF4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rs13039863</td>\n",
       "      <td>HNF4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rs6103716</td>\n",
       "      <td>HNF4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>rs4810426</td>\n",
       "      <td>HNF4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>rs6031563</td>\n",
       "      <td>HNF4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>rs6065725</td>\n",
       "      <td>HNF4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>rs1800961</td>\n",
       "      <td>HNF4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>rs1516728</td>\n",
       "      <td>ETV5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>rs7787720</td>\n",
       "      <td>ETV1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>rs12154701</td>\n",
       "      <td>ETV1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>rs56805921</td>\n",
       "      <td>ETV1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>rs1019029</td>\n",
       "      <td>ETV1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>rs1122518</td>\n",
       "      <td>ETV1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>rs7795991</td>\n",
       "      <td>ETV1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>rs3802219</td>\n",
       "      <td>TRPS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>rs2737226</td>\n",
       "      <td>TRPS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>rs800909</td>\n",
       "      <td>TRPS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>rs79103584</td>\n",
       "      <td>GLIS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>rs510807</td>\n",
       "      <td>GLIS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>rs1574285</td>\n",
       "      <td>GLIS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>rs34494309</td>\n",
       "      <td>GLIS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>rs7041847</td>\n",
       "      <td>GLIS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>rs7034200</td>\n",
       "      <td>GLIS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>rs4237150</td>\n",
       "      <td>GLIS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>rs10974438</td>\n",
       "      <td>GLIS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>rs10758593</td>\n",
       "      <td>GLIS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>rs10814916</td>\n",
       "      <td>GLIS3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>rs4258054</td>\n",
       "      <td>GLIS3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       snp_name   gene\n",
       "0    rs72926932   TCF4\n",
       "1    rs28719468   TCF4\n",
       "2    rs76811102  HNF4A\n",
       "3     rs6017317  HNF4A\n",
       "4     rs4812829  HNF4A\n",
       "5    rs16988991  HNF4A\n",
       "6   rs113810779  HNF4A\n",
       "7    rs12625671  HNF4A\n",
       "8    rs13039863  HNF4A\n",
       "9     rs6103716  HNF4A\n",
       "10    rs4810426  HNF4A\n",
       "11    rs6031563  HNF4A\n",
       "12    rs6065725  HNF4A\n",
       "13    rs1800961  HNF4A\n",
       "14    rs1516728   ETV5\n",
       "15    rs7787720   ETV1\n",
       "16   rs12154701   ETV1\n",
       "17   rs56805921   ETV1\n",
       "18    rs1019029   ETV1\n",
       "19    rs1122518   ETV1\n",
       "20    rs7795991   ETV1\n",
       "21    rs3802219  TRPS1\n",
       "22    rs2737226  TRPS1\n",
       "23     rs800909  TRPS1\n",
       "24   rs79103584  GLIS3\n",
       "25     rs510807  GLIS3\n",
       "26    rs1574285  GLIS3\n",
       "27   rs34494309  GLIS3\n",
       "28    rs7041847  GLIS3\n",
       "29    rs7034200  GLIS3\n",
       "30    rs4237150  GLIS3\n",
       "31   rs10974438  GLIS3\n",
       "32   rs10758593  GLIS3\n",
       "33   rs10814916  GLIS3\n",
       "34    rs4258054  GLIS3"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mapping = pd.DataFrame({'snp_name':[],\n",
    "                          'gene': []})\n",
    "for i in range(len(df_generegions)):\n",
    "    chrom = df_generegions['chrom'][i]\n",
    "    df_gwas_sub = df_gwas_hg19[df_gwas_hg19['chrom']==chrom]\n",
    "    snps = []\n",
    "    tfs = []\n",
    "    for j in range(len(df_gwas_sub)):\n",
    "        if df_gwas_sub['chromEnd'].iloc[j]>df_generegions['start'][i] and df_gwas_sub['chromEnd'].iloc[j]<df_generegions['end'][i]:\n",
    "            snps.append(df_gwas_sub['snp_name'].iloc[j])\n",
    "            tfs.append(df_generegions['gene'][i])\n",
    "    df_temp = pd.DataFrame({'snp_name': snps, 'gene': tfs})\n",
    "    df_mapping = df_mapping.append(df_temp, ignore_index=True)\n",
    "df_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "741c4d75",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mapping.to_csv('/nfs/public/xixi/scRegulate/GWAS/GWAS_SNPs_TFs.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03f09c9e",
   "metadata": {},
   "source": [
    "# pcSNP -> pdCRE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f747ec6b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "239\n"
     ]
    }
   ],
   "source": [
    "files = os.listdir('/nfs/public/xixi/scRegulate/T2D/nn.best.feature6.learnW')\n",
    "markers_filtered = []\n",
    "for file in files:\n",
    "    if len(file.split('.2.pt'))>1:\n",
    "        marker = file.split('.2.pt')[0]\n",
    "        if marker not in markers_filtered:\n",
    "            markers_filtered.append(marker)\n",
    "\n",
    "markers = pd.read_csv(\"/nfs/public/xixi/scRegulate/T2D/data/markers_status.txt\", sep='\\t')\n",
    "markers = markers.groupby('cluster').apply(lambda x: x.nlargest(100, 'avg_log2FC')).reset_index(drop=True)\n",
    "markers = markers['gene'].drop_duplicates().to_list()\n",
    "\n",
    "genes_filtered = [i for i in markers_filtered if i not in tfs_kept]\n",
    "genes_filtered = [i for i in genes_filtered if i in markers]\n",
    "print(len(genes_filtered))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d3c81f08",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1259\n",
      "1015\n",
      "0.8061953931691819\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9</td>\n",
       "      <td>94664770</td>\n",
       "      <td>94665270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr9</td>\n",
       "      <td>94914257</td>\n",
       "      <td>94914757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr9</td>\n",
       "      <td>94671007</td>\n",
       "      <td>94671507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr9</td>\n",
       "      <td>94558673</td>\n",
       "      <td>94559173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9</td>\n",
       "      <td>94660876</td>\n",
       "      <td>94661376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1404</th>\n",
       "      <td>chr5</td>\n",
       "      <td>138145582</td>\n",
       "      <td>138146082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1405</th>\n",
       "      <td>chr6</td>\n",
       "      <td>53950266</td>\n",
       "      <td>53950766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1406</th>\n",
       "      <td>chr5</td>\n",
       "      <td>137872322</td>\n",
       "      <td>137872822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1407</th>\n",
       "      <td>chr8</td>\n",
       "      <td>6473778</td>\n",
       "      <td>6474278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1408</th>\n",
       "      <td>chr5</td>\n",
       "      <td>20575608</td>\n",
       "      <td>20576108</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1409 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         0          1          2\n",
       "0     chr9   94664770   94665270\n",
       "1     chr9   94914257   94914757\n",
       "2     chr9   94671007   94671507\n",
       "3     chr9   94558673   94559173\n",
       "4     chr9   94660876   94661376\n",
       "...    ...        ...        ...\n",
       "1404  chr5  138145582  138146082\n",
       "1405  chr6   53950266   53950766\n",
       "1406  chr5  137872322  137872822\n",
       "1407  chr8    6473778    6474278\n",
       "1408  chr5   20575608   20576108\n",
       "\n",
       "[1409 rows x 3 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "percent = 0.05\n",
    "results = np.load('./predict_status/interpret/input_output_openness_all1_GCN.string.npy')\n",
    "h5f = h5py.File('./predict_status/data_T2D_float16.h5', 'r')\n",
    "peaks = pd.DataFrame(np.array([item.decode('utf-8') for j in h5f['peaks'][:] for item in j]).reshape(-1, 3))\n",
    "h5f.close()\n",
    "import scipy.stats as ss\n",
    "rank_all = []\n",
    "for rep in range(10):\n",
    "    #top = int(0.1*len(tfs_kept))\n",
    "    cells = results[rep, :, :, :]\n",
    "    rank = ss.rankdata(np.array([np.sum(abs(cells[:, i, :])) for i in range(len(peaks))])) #rank from the smallest to the biggest\n",
    "    rank_all.append(rank)\n",
    "\n",
    "rank_all = np.array(rank_all)\n",
    "top = int(percent*len(peaks))\n",
    "top_peaks_ind = np.argsort(np.mean(rank_all, axis=0))[::-1][:top]  # rank from the biggest to the smallest\n",
    "\n",
    "results = np.load('./predict_status/interpret/input_output_openness*0_GCN.string.npy')\n",
    "#results_orig = np.load('./predict_status/interpret/input_output_openness*0_orig_GCN.string.npy')\n",
    "#results_perturb = np.load('./predict_status/interpret/input_output_openness*0_perturb_GCN.string.npy')\n",
    "rank_all = []\n",
    "for rep in range(10):\n",
    "    #top = int(0.1*len(tfs_kept))\n",
    "    cells = results[rep, :, :, :]\n",
    "    rank = ss.rankdata(np.array([np.sum(abs(cells[:, i, :])) for i in range(len(peaks))])) #rank from the smallest to the biggest\n",
    "    rank_all.append(rank)\n",
    "\n",
    "rank_all = np.array(rank_all)\n",
    "top = int(percent*len(peaks))\n",
    "top_peaks_ind0 = np.argsort(np.mean(rank_all, axis=0))[::-1][:top]  # rank from the biggest to the smallest\n",
    "print(top)\n",
    "print(len(set(top_peaks_ind).intersection(set(top_peaks_ind0))))\n",
    "print(len(set(top_peaks_ind).intersection(set(top_peaks_ind0)))/top)\n",
    "\n",
    "top_peaks_ind = list(top_peaks_ind)+list(top_peaks_ind0)\n",
    "top_peaks_ind = list(pd.DataFrame(top_peaks_ind).drop_duplicates()[0])\n",
    "\n",
    "marker_peaks = []\n",
    "markers_ = []\n",
    "_peaks = []\n",
    "df_bed = pd.DataFrame()\n",
    "for i in top_peaks_ind:\n",
    "    for marker_ind in range(len(cut)):\n",
    "        if i>=cut[marker_ind] and i<cut[marker_ind+1]:\n",
    "            break\n",
    "    marker = genes_filtered[marker_ind]\n",
    "    markers_.append(marker)\n",
    "    df_bed = df_bed.append(peaks.iloc[i, :])\n",
    "    peak = str(peaks.iloc[i, 0])+':'+str(peaks.iloc[i, 1])+'-'+str(peaks.iloc[i, 2])\n",
    "    \n",
    "    _peaks.append(peak)\n",
    "    feature = marker + '::' + peak\n",
    "    marker_peaks.append(feature)\n",
    "df_bed = df_bed.drop_duplicates().reset_index(drop=True)\n",
    "df_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c9dcd542",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>peak</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ROR2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>94664770</td>\n",
       "      <td>94665270</td>\n",
       "      <td>ROR2::chr9:94664770-94665270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ROR2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>94914257</td>\n",
       "      <td>94914757</td>\n",
       "      <td>ROR2::chr9:94914257-94914757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IARS</td>\n",
       "      <td>chr9</td>\n",
       "      <td>94914257</td>\n",
       "      <td>94914757</td>\n",
       "      <td>IARS::chr9:94914257-94914757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ROR2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>94671007</td>\n",
       "      <td>94671507</td>\n",
       "      <td>ROR2::chr9:94671007-94671507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ROR2</td>\n",
       "      <td>chr9</td>\n",
       "      <td>94558673</td>\n",
       "      <td>94559173</td>\n",
       "      <td>ROR2::chr9:94558673-94559173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1498</th>\n",
       "      <td>HSPA9</td>\n",
       "      <td>chr5</td>\n",
       "      <td>138145582</td>\n",
       "      <td>138146082</td>\n",
       "      <td>HSPA9::chr5:138145582-138146082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1499</th>\n",
       "      <td>MLIP</td>\n",
       "      <td>chr6</td>\n",
       "      <td>53950266</td>\n",
       "      <td>53950766</td>\n",
       "      <td>MLIP::chr6:53950266-53950766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1500</th>\n",
       "      <td>HSPA9</td>\n",
       "      <td>chr5</td>\n",
       "      <td>137872322</td>\n",
       "      <td>137872822</td>\n",
       "      <td>HSPA9::chr5:137872322-137872822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1501</th>\n",
       "      <td>AGPAT5</td>\n",
       "      <td>chr8</td>\n",
       "      <td>6473778</td>\n",
       "      <td>6474278</td>\n",
       "      <td>AGPAT5::chr8:6473778-6474278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1502</th>\n",
       "      <td>CDH18</td>\n",
       "      <td>chr5</td>\n",
       "      <td>20575608</td>\n",
       "      <td>20576108</td>\n",
       "      <td>CDH18::chr5:20575608-20576108</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1503 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene chrom      start        end                             peak\n",
       "0       ROR2  chr9   94664770   94665270     ROR2::chr9:94664770-94665270\n",
       "1       ROR2  chr9   94914257   94914757     ROR2::chr9:94914257-94914757\n",
       "2       IARS  chr9   94914257   94914757     IARS::chr9:94914257-94914757\n",
       "3       ROR2  chr9   94671007   94671507     ROR2::chr9:94671007-94671507\n",
       "4       ROR2  chr9   94558673   94559173     ROR2::chr9:94558673-94559173\n",
       "...      ...   ...        ...        ...                              ...\n",
       "1498   HSPA9  chr5  138145582  138146082  HSPA9::chr5:138145582-138146082\n",
       "1499    MLIP  chr6   53950266   53950766     MLIP::chr6:53950266-53950766\n",
       "1500   HSPA9  chr5  137872322  137872822  HSPA9::chr5:137872322-137872822\n",
       "1501  AGPAT5  chr8    6473778    6474278     AGPAT5::chr8:6473778-6474278\n",
       "1502   CDH18  chr5   20575608   20576108    CDH18::chr5:20575608-20576108\n",
       "\n",
       "[1503 rows x 5 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks_selected = marker_peaks\n",
    "split_elements = [item.split(\"::\") + item.split(\":\") for item in peaks_selected]\n",
    "split_elements = [[item[0], item[4], item[5]] for item in split_elements]\n",
    "\n",
    "df_peaks_selected = pd.DataFrame(split_elements, columns=[\"gene\", \"chrom\", \"startend\"])\n",
    "df_peaks_selected[['start', 'end']] = df_peaks_selected['startend'].str.split('-', expand=True)\n",
    "df_peaks_selected = df_peaks_selected.drop('startend', axis=1)\n",
    "df_peaks_selected['peak'] = peaks_selected\n",
    "df_peaks_selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6cc53714",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>snp_name</th>\n",
       "      <th>ref</th>\n",
       "      <th>altCount</th>\n",
       "      <th>alts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>183004333</td>\n",
       "      <td>183004334</td>\n",
       "      <td>rs4129858</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>G,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr17</td>\n",
       "      <td>44083947</td>\n",
       "      <td>44083948</td>\n",
       "      <td>rs8067056</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "      <td>C,G,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>51438364</td>\n",
       "      <td>51438365</td>\n",
       "      <td>rs3176466</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>51457199</td>\n",
       "      <td>51457200</td>\n",
       "      <td>rs72906810</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "      <td>C,G,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>51506885</td>\n",
       "      <td>51506886</td>\n",
       "      <td>rs12088739</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>G,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2436</th>\n",
       "      <td>chr4</td>\n",
       "      <td>941517</td>\n",
       "      <td>941518</td>\n",
       "      <td>rs2290402</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2437</th>\n",
       "      <td>chr5</td>\n",
       "      <td>36257017</td>\n",
       "      <td>36257018</td>\n",
       "      <td>rs16902871</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>G,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2438</th>\n",
       "      <td>chr11</td>\n",
       "      <td>64100775</td>\n",
       "      <td>64100776</td>\n",
       "      <td>rs1662185</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>G,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2439</th>\n",
       "      <td>chr3</td>\n",
       "      <td>72803589</td>\n",
       "      <td>72803590</td>\n",
       "      <td>rs9814945</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "      <td>A,T,</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2440</th>\n",
       "      <td>chr3</td>\n",
       "      <td>72813581</td>\n",
       "      <td>72813582</td>\n",
       "      <td>rs11922794</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>A,C,</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2441 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom  chromStart   chromEnd    snp_name ref  altCount  alts\n",
       "0      chr1   183004333  183004334   rs4129858   A         1    G,\n",
       "1     chr17    44083947   44083948   rs8067056   T         2  C,G,\n",
       "2      chr1    51438364   51438365   rs3176466   C         1    T,\n",
       "3      chr1    51457199   51457200  rs72906810   A         2  C,G,\n",
       "4      chr1    51506885   51506886  rs12088739   A         1    G,\n",
       "...     ...         ...        ...         ...  ..       ...   ...\n",
       "2436   chr4      941517     941518   rs2290402   C         1    T,\n",
       "2437   chr5    36257017   36257018  rs16902871   A         1    G,\n",
       "2438  chr11    64100775   64100776   rs1662185   A         1    G,\n",
       "2439   chr3    72803589   72803590   rs9814945   C         2  A,T,\n",
       "2440   chr3    72813581   72813582  rs11922794   G         2  A,C,\n",
       "\n",
       "[2441 rows x 7 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gwas_hg19 = pd.read_csv('/nfs/public/xixi/scRegulate/GWAS/GWAS_T2D_hg19_UCSC.csv')\n",
    "df_gwas_hg19 = df_gwas_hg19.iloc[:, :7]\n",
    "df_gwas_hg19 = df_gwas_hg19.rename(columns={df_gwas_hg19.columns[3]: 'snp_name'})\n",
    "df_gwas_hg19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9929c79c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_name</th>\n",
       "      <th>peak</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs4655617</td>\n",
       "      <td>SGIP1::chr1:67010359-67010859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs112667817</td>\n",
       "      <td>HSPA9::chr5:137822676-137823176</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      snp_name                             peak\n",
       "0    rs4655617    SGIP1::chr1:67010359-67010859\n",
       "1  rs112667817  HSPA9::chr5:137822676-137823176"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mapping = pd.DataFrame({'snp_name': [], 'peak': []})\n",
    "for i in range(len(df_peaks_selected)):\n",
    "    chrom = df_peaks_selected['chrom'][i]\n",
    "    df_gwas_sub = df_gwas_hg19[df_gwas_hg19['chrom']==chrom]\n",
    "    snps = []\n",
    "    peaks = []\n",
    "    for j in range(len(df_gwas_sub)):\n",
    "        if df_gwas_sub['chromEnd'].iloc[j]>int(df_peaks_selected['start'][i]) and df_gwas_sub['chromEnd'].iloc[j]<int(df_peaks_selected['end'][i]):\n",
    "            snps.append(df_gwas_sub['snp_name'].iloc[j])\n",
    "            peaks.append(df_peaks_selected['peak'][i])\n",
    "    df_temp = pd.DataFrame({'snp_name': snps, 'peak': peaks})\n",
    "    df_mapping = df_mapping.append(df_temp, ignore_index=True)\n",
    "df_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e149be7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mapping.to_csv('/nfs/public/xixi/scRegulate/GWAS/GWAS_SNPs_cCREs.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyenv",
   "language": "python",
   "name": "pyenv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
